---
title: "Introduction to microbiomeutilities"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{Introduction to microbiomeutilities}
  %\VignetteEngine{knitr::rmarkdown}
  \usepackage[utf8]{inputenc}
---
  
```{r setup, include = FALSE}
knitr::opts_chunk$set(
collapse = TRUE,
comment = "#>"
)
```

# microbiomeutilties R package  

The microbiomeutilities R package is mainly a wrapper tool with diverse functions for data handling and visualisation. 

## Install  

```{r, eval=FALSE}

install.packages("devtools")
devtools::install_github("microsud/microbiomeutilities")
```  


```{r, message=FALSE, warning=FALSE}

library(microbiomeutilities)
library(microbiome)
library(knitr)
library(tibble)
library(dplyr)
```


## Example data 

Package data from [Zackular et al., 2014: The Gut Microbiome Modulates Colon Tumorigenesis](http://mbio.asm.org/content/4/6/e00692-13).  

Useful resources:  

For more information on phyloseq data structure and uses you can have a look at [Phyloseq](http://joey711.github.io/phyloseq/)  

Tools for microbiome analysis in R. **Microbiome package** URL: [microbiome package](http://microbiome.github.com/microbiome).  


```{r}

data("zackular2014")
ps0 <- zackular2014
ps0
```
The `print_ps` from `microbiomeutilities` can give additional information.  
```{r}
print_ps(ps0)
```

## Microbiome analysis pipeline  

Function `microbiome_pipeline` generates an `HTML` report with preliminary QC, Alpha Diversity, Ordination and Composition analysis of OTU tables. This function saves all intermediate files incuding figures and phyloseq objects in user specified directory.

```{r, eval=FALSE}

microbiome_pipeline(
  otufile = "my.biom",
  mapping = "mymap.csv",
  taxonomy = NULL,
  treefilename = "myTree.tre",
  type = "biom",
  work_dir = "F:/path/my/input/filefolder",
  out_dir = "F:/path/to/save/my/files/folder",
  VariableA = "MC_type1",
  VariableB = "Region",
  UnConstOrd = TRUE,
  heatmap = TRUE,
  filterCount = 4,
  filterPrev = 0.01,
  col.palette = "Paired",
  filterpseq = TRUE,
  samsize = NA,
  projectname = "Mock",
  author = "Sudarshan"
)
```

## Formatting the Phyloseq Object  

Most commonly it is observed that the taxonomy file has classification until a given taxonomic level.We can change the names in both otu table and taxonomy table with the best taxonomic classification available. This can be useful if the analysis has to be done at OTU level. Only ID are less useful.    
  
Check the taxonomy in phyloseq object.  

```{r}

kable(head(tax_table(ps0)))
```

Some have only g__ of s__ information.  

```{r, eval=FALSE}

data("zackular2014")
p0 <- zackular2014
# reduce size for example
ps0 <- core(ps0, detection = 10, prevalence = 20 / 100)
# Add a prefix to taxa labels
ps0.f2 <- format_to_besthit(ps0, prefix = "MyBug-")

kable(head(tax_table(ps0.f2))[3:6])
```

Now the available taxonomy is added.  

As can be seen, the rownames have the OTUIDs and available toxonomic name(s).  

## Summarize the percent taxa classification for phyloseq  

This can be useful to get an overview of taxonomic classifications. Only patterns such as [g__] or `NA` is expected. [g__<empty>] or similar are not considered. Pease convert for eg. g__unclassified to uniform [g__] or simply `NA`.  
This best works with cleaned taxonomy table.  

```{r eval=FALSE}
data("zackular2014")
p0 <- zackular2014
# reduce size for example
ps0 <- core(p0, detection = 10, prevalence = 50 / 100)

percent_classified(ps0)

```


## Distribution of reads  

Useful for QC purposes. Check for siatribution of sequencing depth.     

```{r, eval=TRUE}
data("zackular2014")
p0 <- zackular2014
p <- plot_read_distribution(p0, groups = "DiseaseState", 
                            plot.type = "density")
print(p)
```


## Convert phyloseq object to long data format  

Useful if the user wants to plot specific features.  

```{r}
data("zackular2014")
p0 <- zackular2014
# reduce size for example
ps0 <- core(p0, detection = 10, prevalence = 20 / 100)

pseq_df <- phy_to_ldf(ps0, transform.counts = NULL)

kable(head(pseq_df))
```

## Check distirbution   

A quick check to see how different taxa are distributed in your data.  
```{r message=FALSE, fig.height=6, fig.width=8}
library(microbiomeutilities)
library(ggpubr)
data("zackular2014")
pseq <- zackular2014

# check healthy
health_ps <- subset_samples(pseq, DiseaseState="H")
p_hc <- taxa_distribution(health_ps)

# check CRC
crc_ps <- subset_samples(pseq, DiseaseState="CRC")
p_crc <- taxa_distribution(crc_ps)
ggarrange(p_hc, p_crc, nrow=2, ncol=1)
```

There are Cyanobacteria/Chloroplast related sequences which can be removed if not expected in the samples.  

## Plot alpha diversities  

Utility plot function for diversity measures calculated by `microbiome` package.  

```{r, fig.width=8, fig.height=6, message=FALSE}

library(microbiome)
data("zackular2014")
ps0 <- zackular2014
mycols <- c("brown3", "steelblue","grey50")
p <- plot_alpha_diversities(ps0,
                            type = "dominance",
                            index.val = "all",
                            plot.type = "stripchart",
                            variableA = "DiseaseState",
                            palette = mycols)

print(p)

```

```{r fig.width=8, fig.height=6, message=FALSE}
comps <- make_pairs(sample_data(ps0)$DiseaseState)

p <- p + stat_compare_means(
      comparisons = comps,
      label = "p.format",
      tip.length = 0.05,
      method = "wilcox.test")
p
```


Alternatively, one can plot one index at a time with pair-wise stats. 
```{r message=FALSE, fig.width=4, fig.height=4}

library(gghalves)
data("zackular2014")
p0 <- zackular2014
mycols <- c("brown3", "steelblue","grey50")

p.m <- plot_diversity_stats(p0, group = "DiseaseState", 
                            index = "diversity_shannon", 
                            group.order = c("H", "CRC", "nonCRC"), 
                            group.colors = mycols)
p.m + ylab("Shannon Diversity") + xlab("")

```


## Plot ordination and core  

```{r, fig.height=8, fig.width=12, eval=FALSE}

library(microbiomeutilities)
library(RColorBrewer)
data("zackular2014")
p0 <- zackular2014

ps1 <- format_to_besthit(p0)

#ps1 <- subset_samples(ps1, DiseaseState == "H")
ps1 <- prune_taxa(taxa_sums(ps1) > 0, ps1)
prev.thres <- seq(.05, 1, .05)
det.thres <- 10^seq(log10(1e-4), log10(.2), length = 10)
pseq.rel <- microbiome::transform(ps1, "compositional")
# reduce size for example
pseq.rel <- core(pseq.rel, detection = 0.001, prevalence = 20 / 100)

ord.bray <- ordinate(pseq.rel, "NMDS", "bray")

p <- plot_ordiplot_core(pseq.rel, ord.bray,
                        prev.thres, det.thres,
                        min.prevalence = 0.9,
                        color.opt = "DiseaseState", 
                        shape = NULL, Sample = TRUE)

print(p)

```

## Plot taxa boxplot  

Plot relative abundance of top taxa specified by user. 

```{r,fig.width=7, fig.height=3, eval=TRUE, message=FALSE}
library(microbiomeutilities)
library(RColorBrewer)
data("zackular2014")
ps0 <- zackular2014
mycols <- c("brown3", "steelblue", "grey50")
pn <- plot_taxa_boxplot(ps0,
                        taxonomic.level = "Family",
                        top.otu = 3, 
                        group = "DiseaseState",
                        add.violin= FALSE,
                        title = "Top three family", 
                        keep.other = FALSE,
                        group.order = c("H","CRC","nonCRC"),
                        group.colors = mycols,
                        dot.size = 1)

print(pn + theme_biome_utils())

```

## Plotting selected taxa  

Using a list of taxa specified by the user for comparisons.  
```{r fig.height=4}
library(microbiome)
library(microbiomeutilities)
library(ggpubr)
data("zackular2014")
p0 <- zackular2014
p0.f <- format_to_besthit(p0)
select.taxa <- c("OTU-d__denovo4:Bacteroides", "OTU-d__denovo2:Bacteroides")

mycols <- c("brown3", "steelblue", "grey50")

p <- plot_listed_taxa(p0.f, select.taxa, 
                      group= "DiseaseState",
                      group.order = c("H","CRC","nonCRC"),
                      group.colors = mycols,
                      add.violin = TRUE,
                      violin.opacity = 0.3,
                      dot.opacity = 0.25,
                      box.opacity = 0.25,
                      panel.arrange= "grid")
print(p + ylab("Relative abundance") + scale_y_continuous(labels = scales::percent))


```

Adding statistical test with `ggpubr::stat_compare_means()`  
```{r fig.height=4}

# If more than two variables
comps <- make_pairs(sample_data(p0.f)$DiseaseState)
print(comps)
p <- p + stat_compare_means(
      comparisons = comps,
      label = "p.format",
      tip.length = 0.05,
      method = "wilcox.test")
p

```

Plot top four Genera    
```{r fig.width=6, fig.height=6, eval=TRUE, message=FALSE}
library(microbiome)
library(microbiomeutilities)
data("zackular2014")
p0 <- zackular2014
p0.f <- aggregate_taxa(p0, "Genus")
top_four <- top_taxa(p0.f, 4)
top_four

mycols <- c("brown3", "steelblue", "grey50")

p <- plot_listed_taxa(p0.f, top_four, 
                      group= "DiseaseState",
                      group.order = c("H","CRC","nonCRC"),
                      group.colors = mycols,
                      add.violin = TRUE,
                      violin.opacity = 0.3,
                      dot.opacity = 0.25,
                      box.opacity = 0.25,
                      panel.arrange= "wrap")

comps <- make_pairs(sample_data(p0.f)$DiseaseState)
p <- p + stat_compare_means(
      comparisons = comps,
      label = "p.format",
      tip.length = 0.05,
      method = "wilcox.test")

print(p + ylab("Relative abundance") + scale_y_continuous(labels = scales::percent))
```


## Abundance-Prevalence relationship  
Plots Mean Abundance-Prevalence for taxa. Mean abundance, mean prevalence, and upper and lower confidence interval for each taxa is calculated by random subsampling.  
This can be useful in identifying highly prevalent taxa and thier mean relative abundance in a group of samples. Taxa that are highly prevalent with low variation in lower and upper CI can be identified at varying values of mean relative abundance. These are likely core taxa in the groups of samples.

```{r fig.height=4, fig.width=6}
library(microbiomeutilities)
library(dplyr)
library(ggrepel)
asv_ps <- zackular2014
asv_ps <- microbiome::transform(asv_ps, "compositional")
# Select healthy samples
asv_ps <- subset_samples(asv_ps, DiseaseState=="H")
asv_ps <- core(asv_ps,detection = 0.0001, prevalence = 0.50) # reduce size for example
asv_ps <- format_to_besthit(asv_ps)

set.seed(14285)
p_v <- plot_abund_prev(asv_ps, 
                       label.core = TRUE,
                       color = NULL,
                       mean.abund.thres = 0.01,
                       mean.prev.thres = 0.99,
                       point.opacity = 0.1,
                       label.size = 3,
                       label.opacity = 1.0,
                       nudge.label=-0.15,
                       bs.iter=9,
                       size = 20, 
                       replace = TRUE)  
p_v <- p_v + 
  geom_vline(xintercept = 0.95, lty="dashed", alpha=0.7) + 
  geom_hline(yintercept = 0.01,lty="dashed", alpha=0.7) +
  scale_color_brewer(palette = "Dark2")
p_v
```

## Heatmaps    
### Heatmap using phyloseq and pheatmap  
Useful for visualisng differences in top otus between sample groups.  
```{r fig.height=6, fig.width=8}
library(pheatmap)
library(RColorBrewer)
data("zackular2014")
ps0 <- zackular2014

plot_taxa_heatmap(ps0,
                  subset.top = 20,
                  VariableA = "DiseaseState",
                  heatcolors = rev(brewer.pal(6, "Spectral")),
                  transformation = "clr",
                  cluster_rows = F,
                  cluster_cols = F,
                  show_colnames = F)
```

### Elegant option with ggplot2    

```{r fig.width=10, fig.height=6, message=FALSE}
library(microbiomeutilities)
data("zackular2014")
p0 <- zackular2014
p0.rel <- transform(p0, "compositional")
simple_heatmap(p0.rel,
               group.facet = "DiseaseState",
               group.order = c("H", "CRC", "nonCRC"),
               abund.thres = 0.01,
               prev.thres = 0.1,
               level = "Genus",
               scale.color = "log10",
               na.fill = "white",
               color.fill = NULL,
               taxa.arrange=NULL,
               remove.other=TRUE,
               panel.arrange="grid",
               ncol=NULL,
               nrow=NULL)
```

## MicrobiomeHD datasets as phyloseq objects  

We provide access to a subset of studies included in the `MicrobiomeHD` database from Duvallet et al 2017: [Meta-analysis of gut microbiome studies identifies disease-specific and shared responses](https://www.nature.com/articles/s41467-017-01973-8#ref-CR33). Nature communications.    

The phyloseq objects are stored and accessed from [microbiomedatarepo](https://github.com/microsud/microbiomedatarepo).  

```{r}

study <- list_microbiome_data(printtab = FALSE)

knitr::kable(study)
```


Below is the per study reference.  

**NOTE**: When using these studies, please cite [Duvallet et al. 2017](https://www.nature.com/articles/s41467-017-01973-8#ref-CR33) and the respective studies.  

```{r, eval=FALSE}

file <- system.file("extdata", "microbiomeHD_ref.txt", package = "microbiomeutilities")
reference <- read.table(file, header = T, sep = "\t")

knitr::kable(reference)
```

For more tutorials and examples of data anlaysis in R please check:  

* [Microbiome data analysis SpringSchool2018](https://goo.gl/CPChhd)  
* [Microbiome R package tutorials](http://microbiome.github.io/microbiome/)  

```{r}

sessionInfo()
```


## Package requirements  

* Depends:
+ [R (>= 3.4.1)](https://www.r-project.org/)
+ [phyloseq](https://joey711.github.io/phyloseq/index.html)
+ [ggplot2](http://ggplot2.org/)
+ [microbiome](https://bioconductor.org/packages/devel/bioc/html/microbiome.html)

* Imports:
+ [plyr](https://cran.r-project.org/web/packages/plyr/index.html)
+ [reshape2](https://cran.rstudio.com/web/packages/reshape2/index.html)
+ [stats](https://stat.ethz.ch/R-manual/R-devel/library/stats/html/00Index.html)
+ [tidyr](http://tidyr.tidyverse.org/)
+ [utils](https://stat.ethz.ch/R-manual/R-devel/library/utils/html/00Index.html)
+ [vegan](https://cran.r-project.org/web/packages/vegan/index.html)
+ [ggpubr](https://cran.r-project.org/web/packages/ggpubr/index.html)
+ [RColorBrewer](https://cran.r-project.org/web/packages/RColorBrewer/index.html)
+ [viridis](https://cran.r-project.org/web/packages/viridis/index.html)
+ [pheatmap](https://cran.r-project.org/web/packages/pheatmap/index.html)
+ [ggrepel](https://cran.r-project.org/web/packages/ggrepel/index.html)

* Suggests:
+ [BiocGenerics](https://bioconductor.org/packages/release/bioc/html/BiocGenerics.html)
+ [knitr](https://cran.r-project.org/web/packages/knitr/index.html)
+ [rmarkdown](http://rmarkdown.rstudio.com/)
+ [ape](https://cran.r-project.org/web/packages/ape/index.html)
+ [Picante](https://academic.oup.com/bioinformatics/article/26/11/1463/203321)
