<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>microbiomeutilities: Utilities for Microbiome Analytics • microbiomeutilities</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="bootstrap-toc.css">
<script src="bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="pkgdown.css" rel="stylesheet">
<script src="pkgdown.js"></script><meta property="og:title" content="microbiomeutilities: Utilities for Microbiome Analytics">
<meta property="og:description" content="This is a supporting tool for extending the functionality of both phyloseq and microbiome R packages.">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-home">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="index.html">microbiomeutilities</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">1.00.09</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="articles/microbiomeutilities.html">Get started</a>
</li>
<li>
  <a href="reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="articles/microbiomeutilities_ts.html">Longitudinal data analysis and visualization</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/microsud/microbiomeutilities/">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="contents col-md-9">


<div id="about" class="section level1">
<div class="page-header"><h1 class="hasAnchor">
<a href="#about" class="anchor"></a>About</h1></div>
<p>The <code>microbiomeutilities</code> R package is part of the <a href="https://github.com/microbiome">microbiome-verse</a> tools that provides additional data handling and visualization support for the <a href="https://microbiome.github.io/tutorials/">microbiome R/BioC package</a></p>
<p>Philosophy: “<em>Seemingly simple tasks for experienced R users can always be further simplified for novice users</em>”</p>
<p><a href="https://microsud.github.io/microbiomeutilities/">Package website and online documentation</a></p>
<div id="install-microbiomeutilities" class="section level2">
<h2 class="hasAnchor">
<a href="#install-microbiomeutilities" class="anchor"></a>Install microbiomeutilities</h2>
<pre><code><a href="https://rdrr.io/r/utils/install.packages.html">install.packages("devtools")
devtools::install_github("microsud/microbiomeutilities")
</a></code></pre>
<p><strong>Citation:</strong></p>
<p>o Leo Lahti, Sudarshan Shetty et al. (2017-2020). Tools for microbiome analysis in R. Version 2.1.28. URL: <a href="http://microbiome.github.com/microbiome" class="uri">http://microbiome.github.com/microbiome</a><br>
o Sudarshan A. Shetty, &amp; Leo Lahti. (2020, October). <code>microbiomeutilities</code>: Utilities for Microbiome Analytics (Version 1.00.09).</p>
<p>The microbiome R package relies on the independently developed<br>
o <a href="https://github.com/joey711/phyloseq">phyloseq</a> package and data structures for R-based microbiome analysis developed by Paul McMurdie and Susan Holmes.<br>
o <a href="http://ggplot2.org/">ggplot2</a> H. Wickham. ggplot2: Elegant Graphics for Data Analysis. Springer-Verlag New York, 2009.<br>
o <a href="https://www.tidyverse.org/">tidyverse packages</a>.</p>
<p><strong>Microbiome package website with step-wise tutorials</strong>:<br>
URL: <a href="http://microbiome.github.com/microbiome">http://microbiome.github.com/microbiome</a>.</p>
<p><strong>Tutorials</strong></p>
<ul>
<li>
<a href="http://microbiome.github.io/microbiome/">Microbiome R package tutorials</a><br>
</li>
<li>
<a href="https://mibwurrepo.github.io/Microbial-bioinformatics-introductory-course-Material-2018/introduction.html">Open and reproducible spring school 2018</a><br>
</li>
<li><a href="https://microsud.github.io/Tools-Microbiome-Analysis/">Tools Microbiome Anlaysis</a></li>
</ul>
<div id="about-the-author" class="section level3">
<h3 class="hasAnchor">
<a href="#about-the-author" class="anchor"></a>About the Author</h3>
<ul>
<li>
<a href="https://microsud.github.io/gutmicrobe_webpage.io/">Web CV</a><br>
</li>
<li>
<a href="https://twitter.com/gutmicrobe">twitter</a><br>
</li>
<li>
<a href="https://scholar.google.nl/citations?hl=en&amp;user=Vahc6LUAAAAJ&amp;view_op=list_works&amp;sortby=pubdate">googlescholar</a><br>
</li>
<li><a href="http://orcid.org/0000-0001-7280-9915">ORCID ID: 0000-0001-7280-9915</a></li>
</ul>
<p><strong>MicrobiomeHD</strong><br>
The package provides access to a subset of studies included in the <a href="https://zenodo.org/record/840333#.WuYuDoiFM2w">MicrobiomeHD</a> database from Duvallet et al 2017: <a href="https://www.nature.com/articles/s41467-017-01973-8">Meta-analysis of gut microbiome studies identifies disease-specific and shared responses. Nature communications</a>. These datasets are converted to phyloseq objects and can be directly used in R environment.</p>
</div>
<div id="datasets-from" class="section level3">
<h3 class="hasAnchor">
<a href="#datasets-from" class="anchor"></a>Datasets from:</h3>
<ul>
<li>Duvallet, Claire, et al. “Meta-analysis of gut microbiome studies identifies disease-specific and shared responses.” Nature communications 8.1 (2017): 1784.<br>
</li>
<li>Son, J. et al. Comparison of fecal microbiota in children with autism spectrum disorders and neurotypical siblings in the simons simplex collection. PLoS ONE 10, e0137725 (2015).<br>
</li>
<li>Kang, D. W. et al. Reduced incidence of Prevotella and other fermenters in intestinal microflora of autistic children. PLoS ONE8, e68322 (2013).<br>
</li>
<li>Schubert, A. M. et al. Microbiome data distinguish patients with clostridium difficile infection and non-c. difficile-associated diarrhea from healthy controls. mBio 5, e01021–14–e01021–14 (2014).<br>
</li>
<li>Youngster, I. et al. Fecal microbiota transplant for relapsing clostridium difficile infection using a frozen inoculum from unrelated donors: a randomized, open-label, controlled pilot study. Clin. Infect. Dis. 58, 1515–1522 (2014).<br>
</li>
<li>Baxter, N. T., Ruffin, M. T., Rogers, M. A. &amp; Schloss, P. D. Microbiota-based model improves the sensitivity of fecal immunochemical test for detecting colonic lesions. Genome Med. 8, 37 (2016).<br>
</li>
<li>Zackular, Joseph P., et al. “The gut microbiome modulates colon tumorigenesis.” MBio 4.6 (2013): e00692-13.</li>
<li>Zeller, G. et al. Potential of fecal microbiota for early-stage detection of colorectal cancer. Mol. Syst. Biol. 10, 766–766 (2014).</li>
<li>Singh, P. et al. Intestinal microbial communities associated with acute enteric infections and disease recovery. Microbiome 3, 45 (2015).<br>
</li>
<li>Noguera-Julian, M. et al. Gut microbiota linked to sexual preference and hiv infection. EBioMedicine 5, 135–146 (2016). Dinh, D. M. et al. Intestinal microbiota, microbial translocation, and systemic inflammation in chronic HIV infection. J. Infect. Dis. 211, 19–27 (2014).<br>
</li>
<li>Lozupone, C. A. et al. Alterations in the gut microbiota associated with hiv-1 infection. Cell Host Microbe 14, 329–339 (2013).</li>
<li>Gevers, D. et al. The treatment-naive microbiome in new-onset crohn’s disease. Cell Host Microbe 15, 382–392 (2014).</li>
<li>Zhang, Z. et al Large-scale survey of gut microbiota associated with MHE via 16s rRNA-based pyrosequencing. Am. J. Gastroenterol. 108, 1601–1611 (2013).<br>
</li>
<li>Wong, J. M. W., Souza, R. De, Kendall, C. W. C., Emam, A. &amp; Jenkins, D. J. A. Colonic health: fermentation and short chain fatty acids. J. Clin. Gastroenterol. 40, 235–243 (2006).<br>
</li>
<li>Ross, M. C. et al. 16s gut community of the cameron county hispanic cohort. Microbiome 3, 7 (2015).</li>
<li>Zupancic, M. L. et al. Analysis of the gut microbiota in the old order Amish and its relation to the metabolic syndrome. PLoS ONE 7, e43052 (2012).<br>
</li>
<li>Scher, J. U. et al. Expansion of intestinal Prevotella copri correlates with enhanced susceptibility to arthritis. eLife 2, e01202 (2013).<br>
</li>
<li>Alkanani, A. K. et al. Alterations in intestinal microbiota correlate with susceptibility to type 1 diabetes. Diabetes 64, 3510–3520 (2015).<br>
</li>
<li>Scheperjans, F. et al Gut microbiota are related to parkinson’s disease and clinical phenotype. Mov. Disord. 30, 350–358 (2014).</li>
</ul>
<p><strong>NOTE:</strong><br>
The aim of this package is not to replace any of the other tools mentioned on this site.</p>
</div>
</div>
<div id="change-log" class="section level2">
<h2 class="hasAnchor">
<a href="#change-log" class="anchor"></a>Change log</h2>
<p>CHANGES IN VERSION 1.00.09 (2020-10-21)<br>
o Added longitudinal page to website<br>
o Added new function <code>plot_area</code><br>
o Added new function <code>plot_paired_abundances</code><br>
o Added new function <code>plot_spaghetti</code></p>
<p>CHANGES IN VERSION 1.00.08 (2020-10-17)<br>
o Cosmetic updates</p>
<p>CHANGES IN VERSION 1.00.07 (2020-10-15)<br>
o Update to <code>plot_taxa_heatmap</code><br>
o Fixed <code>plot_abund_prev</code> options<br>
o Added <code>plot_alpha_rcurve</code><br>
CHANGES IN VERSION 1.00.06 (2020-10-13)<br>
o Added new function dominant_taxa<br>
o Removed plot_ternary due to clash between ggplot2 and ggtern<br>
o Added new function find_samples_taxa</p>
<p>CHANGES IN VERSION 1.00.05 (2020-10-11)<br>
o Added new function <code>dominant_taxa</code><br>
o Added new function <code>get_group_abundances</code><br>
o removed <code>microbiome_pipeline</code></p>
<p>CHANGES IN VERSION 1.00.04 (2020-10-11)<br>
o Added new function <code>plasticity</code><br>
o modified <code>theme_biome_utils</code></p>
<p>CHANGES IN VERSION 1.00.03 (2020-10-04)<br>
o Added new function plot_ternary<br>
o Deprecated plot_select_taxa</p>
<p>CHANGES IN VERSION 1.00.02 (2020-10-04)<br>
o Version tested with R version 4.0.2 (2020-10-04)<br>
o Added new function <code>plot_listed_taxa</code><br>
o Deprecated <code>plot_select_taxa</code><br>
o Added option for half violin in boxplots</p>
<p>CHANGES IN VERSION 1.00.01 (2020-10-03) o Added new function <code>plot_abund_prev</code><br>
o Added new function <code>simple_heatmap</code><br>
o Added new function <code>taxa_distribution</code><br>
o Added a custom theme <code>theme_biome_utils</code><br>
o Added <code>gghalves</code> to imports<br>
o Fixed <code>microbiome_pipeline</code> report</p>
<p>CHANGES IN VERSION 1.00.00 (2020-10-01) o Version tested with R version 4.0.2 (2020-06-22)<br>
o Fix typos in documentation<br>
o Add prefix option to <code>format_to_besthit</code><br>
o Edited <code>phy_to_ldf</code> to speed up conversion<br>
o Removed <code>format_phyloseq</code> function as it is redundant<br>
o Speedup <code>taxa_summary</code> function<br>
o Free up <code>pheatmap</code> option in <code>plot_taxa_heatmap</code><br>
o Updated for more info from <code>print_ps</code> output<br>
o <code>plot_taxa_boxplot</code> now returns a faceted plot<br>
o Added new function <code>plot_diversity_stats</code><br>
o R code styling <code><a href="https://rdrr.io/pkg/styler/man/tidyverse_style.html">styler::tidyverse_style()</a></code></p>
<div id="references" class="section level3">
<h3 class="hasAnchor">
<a href="#references" class="anchor"></a>References:</h3>
<ol>
<li>Callahan, B. J., McMurdie, P. J. &amp; Holmes, S. P. (2017). Exact sequence variants should replace operational taxonomic units in marker gene data analysis. bioRxiv, 113597.<br>
</li>
<li>Callahan, B. J., McMurdie, P. J., Rosen, M. J., Han, A. W., Johnson, A. J. A. &amp; Holmes, S. P. (2016). DADA2: high-resolution sample inference from Illumina amplicon data. Nature methods 13, 581-583.<br>
</li>
<li>Caporaso, J. G., Kuczynski, J., Stombaugh, J., Bittinger, K., Bushman, F. D., Costello, E. K., Fierer, N., Peña, A. G., Goodrich, J. K. &amp; Gordon, J. I. (2010). QIIME allows analysis of high-throughput community sequencing data. Nature methods 7, 335-336.<br>
</li>
<li>Schloss, P. D., Westcott, S. L., Ryabin, T., Hall, J. R., Hartmann, M., Hollister, E. B., Lesniewski, R. A., Oakley, B. B., Parks, D. H. &amp; Robinson, C. J. (2009). Introducing mothur: open-source, platform-independent, community-supported software for describing and comparing microbial communities. Applied and environmental microbiology 75, 7537-7541.<br>
Team, R. C. (2000). R language definition. Vienna, Austria: R foundation for statistical computing.</li>
</ol>
</div>
<div id="more-useful-resources" class="section level3">
<h3 class="hasAnchor">
<a href="#more-useful-resources" class="anchor"></a>More useful resources:</h3>
<ol>
<li>
<a href="https://f1000research.com/articles/5-1492/v2">Ben J. Callahan and Colleagues: Bioconductor Workflow for Microbiome Data Analysis: from raw reads to community analyses</a>.<br>
</li>
<li>
<a href="http://msystems.asm.org/content/2/1/e00127-16">Comeau AM and Colleagues: Microbiome Helper: a Custom and Streamlined Workflow for Microbiome Research</a><br>
</li>
<li>MicrobiomeHD <a href="http://www.biorxiv.org/content/early/2017/05/08/134031">A standardized database of human gut microbiome studies in health and disease <em>Case-Control</em></a><br>
</li>
<li>Rhea <a href="https://peerj.com/articles/2836/">A pipeline with modular R scripts</a><br>
</li>
<li>Phyloseq <a href="http://journals.plos.org/plosone/article?id=10.1371/journal.pone.0061217">Import, share, and analyze microbiome census data using R</a>
</li>
</ol>
</div>
</div>
</div>

  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <div class="links">
<h2>Links</h2>
<ul class="list-unstyled">
<li>Browse source code at <br><a href="https://github.com/microsud/microbiomeutilities/">https://​github.com/​microsud/​microbiomeutilities/​</a>
</li>
<li>Report a bug at <br><a href="https://github.com/microsud/microbiomeutilities/issues">https://​github.com/​microsud/​microbiomeutilities/​issues</a>
</li>
</ul>
</div>
<div class="license">
<h2>License</h2>
<ul class="list-unstyled">
<li>
<a href="https://opensource.org/licenses/BSD-2-Clause">BSD_2_clause</a> + file <a href="LICENSE-text.html">LICENSE</a> </li>
</ul>
</div>
<div class="developers">
<h2>Developers</h2>
<ul class="list-unstyled">
<li>Sudarshan Shetty <br><small class="roles"> Author, maintainer </small>  </li>
<li>Leo Lahti <br><small class="roles"> Author </small>  </li>
</ul>
</div>

  <div class="dev-status">
<h2>Dev status</h2>
<ul class="list-unstyled">
<li><a href="https://travis-ci.org/microsud/microbiomeutilities"><img src="https://travis-ci.org/microsud/microbiomeutilities.svg?branch=master" alt="Build Status"></a></li>
</ul>
</div>
</div>
</div>


      <footer><div class="copyright">
  <p>Developed by Sudarshan Shetty, Leo Lahti.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
