<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Introduction to microbiomeutilities • microbiomeutilities</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Introduction to microbiomeutilities">
<meta property="og:description" content="microbiomeutilities">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">microbiomeutilities</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.00.17</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../articles/microbiomeutilities.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/microbiomeutilities_ts.html">Longitudinal data analysis and visualization</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/microsud/microbiomeutilities/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Introduction to microbiomeutilities</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/microsud/microbiomeutilities/blob/HEAD/vignettes/microbiomeutilities.Rmd" class="external-link"><code>vignettes/microbiomeutilities.Rmd</code></a></small>
      <div class="hidden name"><code>microbiomeutilities.Rmd</code></div>

    </div>

    
    
<p><strong>NOTE</strong> While we continue to maintain this R package,
the development has been discontinued as we have shifted to supporting
methods development based on the new TreeSummarizedExperiment data
container, which provides added capabilities for multi-omics data
analysis. Check the miaverse project for details.</p>
<p>The <code>microbiomeutilities</code> R package is part of the <a href="https://github.com/microbiome" class="external-link">microbiome-verse</a> tools that
provides additional data handling and visualization support for the <a href="https://microbiome.github.io/tutorials/" class="external-link">microbiome R/BioC
package</a></p>
<p>Philosophy: “<em>Seemingly simple tasks for experienced R users can
always be further simplified for novice users</em>”</p>
<div class="section level2">
<h2 id="install">Install<a class="anchor" aria-label="anchor" href="#install"></a>
</h2>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"devtools"</span><span class="op">)</span></span>
<span><span class="fu">devtools</span><span class="fu">::</span><span class="fu"><a href="https://remotes.r-lib.org/reference/install_github.html" class="external-link">install_github</a></span><span class="op">(</span><span class="st">"microsud/microbiomeutilities"</span><span class="op">)</span></span></code></pre></div>
<p>Load libraries</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">microbiomeutilities</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="http://microbiome.github.io/microbiome" class="external-link">microbiome</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://yihui.org/knitr/" class="external-link">knitr</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://tibble.tidyverse.org/" class="external-link">tibble</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://dplyr.tidyverse.org" class="external-link">dplyr</a></span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="basics">Basics<a class="anchor" aria-label="anchor" href="#basics"></a>
</h2>
<div class="section level3">
<h3 id="example-data">Example data<a class="anchor" aria-label="anchor" href="#example-data"></a>
</h3>
<p>Package data from <a href="http://mbio.asm.org/content/4/6/e00692-13" class="external-link">Zackular et al., 2014:
The Gut Microbiome Modulates Colon Tumorigenesis</a>.</p>
<p>Import test data.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"zackular2014"</span><span class="op">)</span></span>
<span><span class="va">ps0</span> <span class="op">&lt;-</span> <span class="va">zackular2014</span></span>
<span><span class="va">ps0</span></span>
<span><span class="co">#&gt; phyloseq-class experiment-level object</span></span>
<span><span class="co">#&gt; otu_table()   OTU Table:         [ 2078 taxa and 88 samples ]</span></span>
<span><span class="co">#&gt; sample_data() Sample Data:       [ 88 samples by 7 sample variables ]</span></span>
<span><span class="co">#&gt; tax_table()   Taxonomy Table:    [ 2078 taxa by 7 taxonomic ranks ]</span></span></code></pre></div>
<p>The <code>print_ps</code> from <code>microbiomeutilities</code> can
give additional information. See also
<code><a href="https://rdrr.io/pkg/microbiome/man/summarize_phyloseq.html" class="external-link">microbiome::summarize_phyloseq</a></code>.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/print_ps.html">print_ps</a></span><span class="op">(</span><span class="va">ps0</span><span class="op">)</span></span>
<span><span class="co">#&gt; 01] ntaxa = 2078</span></span>
<span><span class="co">#&gt; 02] nsamples = 88</span></span>
<span><span class="co">#&gt; 03] nsamplesvariables = 7</span></span>
<span><span class="co">#&gt; 04] nranks = 7</span></span>
<span><span class="co">#&gt; 05] Min. number of reads = 15704</span></span>
<span><span class="co">#&gt; 06] Max. number of reads = 216888</span></span>
<span><span class="co">#&gt; 07] Total number of reads = 4594626</span></span>
<span><span class="co">#&gt; 08] Average number of reads = 52211.66</span></span>
<span><span class="co">#&gt; 09] Median number of reads = 49415.5</span></span>
<span><span class="co">#&gt; 10] Sparsity = 0.728426590252865</span></span>
<span><span class="co">#&gt; 11] Number of singletons = 0</span></span>
<span><span class="co">#&gt; 12] % of taxa that are singletons </span></span>
<span><span class="co">#&gt;   (i.e. exactly one read detected across all samples) = 0</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="formatting-the-phyloseq-object">Formatting the Phyloseq Object<a class="anchor" aria-label="anchor" href="#formatting-the-phyloseq-object"></a>
</h3>
<p>Most commonly it is observed that the taxonomy file has
classification until a given taxonomic level. We can change the names in
both otu table and taxonomy table with the best taxonomic classification
available. This can be useful if the analysis has to be done at OTU/ASVs
level. Only ID are less useful.</p>
<p>Check the taxonomy in phyloseq object.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/knitr/man/kable.html" class="external-link">kable</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/phyloseq/man/tax_table-methods.html" class="external-link">tax_table</a></span><span class="op">(</span><span class="va">ps0</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<table class="table">
<colgroup>
<col width="11%">
<col width="10%">
<col width="14%">
<col width="12%">
<col width="14%">
<col width="16%">
<col width="12%">
<col width="6%">
</colgroup>
<thead><tr class="header">
<th align="left"></th>
<th align="left">Domain</th>
<th align="left">Phylum</th>
<th align="left">Class</th>
<th align="left">Order</th>
<th align="left">Family</th>
<th align="left">Genus</th>
<th align="left">Species</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left">d__denovo1773</td>
<td align="left">k__Bacteria</td>
<td align="left">p__Bacteroidetes</td>
<td align="left">c__Bacteroidia</td>
<td align="left">o__Bacteroidales</td>
<td align="left">f__Bacteroidaceae</td>
<td align="left">g__Bacteroides</td>
<td align="left">s__</td>
</tr>
<tr class="even">
<td align="left">d__denovo1771</td>
<td align="left">k__Bacteria</td>
<td align="left">p__Bacteroidetes</td>
<td align="left">c__Bacteroidia</td>
<td align="left">o__Bacteroidales</td>
<td align="left">f__Bacteroidaceae</td>
<td align="left">g__Bacteroides</td>
<td align="left">s__</td>
</tr>
<tr class="odd">
<td align="left">d__denovo1776</td>
<td align="left">k__Bacteria</td>
<td align="left">p__Firmicutes</td>
<td align="left">c__Clostridia</td>
<td align="left">o__Clostridiales</td>
<td align="left">f__Ruminococcaceae</td>
<td align="left">g__</td>
<td align="left">s__</td>
</tr>
<tr class="even">
<td align="left">d__denovo1777</td>
<td align="left">k__Bacteria</td>
<td align="left">p__Bacteroidetes</td>
<td align="left">c__Bacteroidia</td>
<td align="left">o__Bacteroidales</td>
<td align="left">f__Bacteroidaceae</td>
<td align="left">g__Bacteroides</td>
<td align="left">s__</td>
</tr>
<tr class="odd">
<td align="left">d__denovo1775</td>
<td align="left">k__Bacteria</td>
<td align="left">p__Firmicutes</td>
<td align="left">c__Clostridia</td>
<td align="left">o__Clostridiales</td>
<td align="left">f__Lachnospiraceae</td>
<td align="left">g__Blautia</td>
<td align="left">s__</td>
</tr>
<tr class="even">
<td align="left">d__denovo2639</td>
<td align="left">k__Bacteria</td>
<td align="left">p__Bacteroidetes</td>
<td align="left">c__Bacteroidia</td>
<td align="left">o__Bacteroidales</td>
<td align="left">f__Bacteroidaceae</td>
<td align="left">g__Bacteroides</td>
<td align="left">s__</td>
</tr>
</tbody>
</table>
<p>Some have only g__ of s__ information.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"zackular2014"</span><span class="op">)</span></span>
<span><span class="va">p0</span> <span class="op">&lt;-</span> <span class="va">zackular2014</span></span>
<span><span class="co"># reduce size for example</span></span>
<span><span class="va">ps0</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/microbiome/man/core.html" class="external-link">core</a></span><span class="op">(</span><span class="va">p0</span>, detection <span class="op">=</span> <span class="fl">10</span>, prevalence <span class="op">=</span> <span class="fl">20</span> <span class="op">/</span> <span class="fl">100</span><span class="op">)</span></span>
<span><span class="co"># Add a prefix to taxa labels</span></span>
<span><span class="va">ps0.f2</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/format_to_besthit.html">format_to_besthit</a></span><span class="op">(</span><span class="va">ps0</span>, prefix <span class="op">=</span> <span class="st">"MyBug-"</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/knitr/man/kable.html" class="external-link">kable</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/phyloseq/man/tax_table-methods.html" class="external-link">tax_table</a></span><span class="op">(</span><span class="va">ps0.f2</span><span class="op">)</span><span class="op">)</span><span class="op">[</span><span class="fl">3</span><span class="op">:</span><span class="fl">6</span><span class="op">]</span><span class="op">)</span></span></code></pre></div>
<table style="width:100%;" class="table">
<colgroup>
<col width="21%">
<col width="5%">
<col width="6%">
<col width="6%">
<col width="7%">
<col width="9%">
<col width="10%">
<col width="10%">
<col width="21%">
</colgroup>
<thead><tr class="header">
<th align="left"></th>
<th align="left">Domain</th>
<th align="left">Phylum</th>
<th align="left">Class</th>
<th align="left">Order</th>
<th align="left">Family</th>
<th align="left">Genus</th>
<th align="left">Species</th>
<th align="left">best_hit</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left">MyBug-d__denovo165:f__Ruminococcaceae</td>
<td align="left">Bacteria</td>
<td align="left">Firmicutes</td>
<td align="left">Clostridia</td>
<td align="left">Clostridiales</td>
<td align="left">Ruminococcaceae</td>
<td align="left">f__Ruminococcaceae</td>
<td align="left">f__Ruminococcaceae</td>
<td align="left">MyBug-d__denovo165:f__Ruminococcaceae</td>
</tr>
<tr class="even">
<td align="left">MyBug-d__denovo167:g__Coprococcus</td>
<td align="left">Bacteria</td>
<td align="left">Firmicutes</td>
<td align="left">Clostridia</td>
<td align="left">Clostridiales</td>
<td align="left">Lachnospiraceae</td>
<td align="left">Coprococcus</td>
<td align="left">g__Coprococcus</td>
<td align="left">MyBug-d__denovo167:g__Coprococcus</td>
</tr>
<tr class="odd">
<td align="left">MyBug-d__denovo166:o__Clostridiales</td>
<td align="left">Bacteria</td>
<td align="left">Firmicutes</td>
<td align="left">Clostridia</td>
<td align="left">Clostridiales</td>
<td align="left">o__Clostridiales</td>
<td align="left">o__Clostridiales</td>
<td align="left">o__Clostridiales</td>
<td align="left">MyBug-d__denovo166:o__Clostridiales</td>
</tr>
<tr class="even">
<td align="left">MyBug-d__denovo161:g__Roseburia</td>
<td align="left">Bacteria</td>
<td align="left">Firmicutes</td>
<td align="left">Clostridia</td>
<td align="left">Clostridiales</td>
<td align="left">Lachnospiraceae</td>
<td align="left">Roseburia</td>
<td align="left">g__Roseburia</td>
<td align="left">MyBug-d__denovo161:g__Roseburia</td>
</tr>
</tbody>
</table>
<p>Now the available taxonomy is added.</p>
<p>As can be seen, the rownames have the OTUIDs and available taxonomic
name(s).</p>
</div>
<div class="section level3">
<h3 id="distribution-of-reads">Distribution of reads<a class="anchor" aria-label="anchor" href="#distribution-of-reads"></a>
</h3>
<p>Useful for QC purposes. Check for distribution of sequencing
depth.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"zackular2014"</span><span class="op">)</span></span>
<span><span class="va">p0</span> <span class="op">&lt;-</span> <span class="va">zackular2014</span></span>
<span><span class="va">p</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/plot_read_distribution.html">plot_read_distribution</a></span><span class="op">(</span><span class="va">p0</span>, groups <span class="op">=</span> <span class="st">"DiseaseState"</span>, </span>
<span>                            plot.type <span class="op">=</span> <span class="st">"density"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">p</span> <span class="op">+</span> <span class="fu"><a href="../reference/theme_biome_utils.html">theme_biome_utils</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="microbiomeutilities_files/figure-html/unnamed-chunk-7-1.png" width="768"></p>
<p>This is a diagnostic step. Key to check if there is variation between
groups that will be compared for downstream analysis.</p>
</div>
<div class="section level3">
<h3 id="convert-phyloseq-object-to-long-data-format">Convert phyloseq object to long data format<a class="anchor" aria-label="anchor" href="#convert-phyloseq-object-to-long-data-format"></a>
</h3>
<p>Useful if the user wants to plot specific features.<br>
Converting to long data format opens several opportunities to work with
<a href="https://www.tidyverse.org/" class="external-link">Tidyverse</a>.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"zackular2014"</span><span class="op">)</span></span>
<span><span class="va">p0</span> <span class="op">&lt;-</span> <span class="va">zackular2014</span></span>
<span><span class="co"># reduce size for example</span></span>
<span><span class="va">ps0</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/microbiome/man/core.html" class="external-link">core</a></span><span class="op">(</span><span class="va">p0</span>, detection <span class="op">=</span> <span class="fl">10</span>, prevalence <span class="op">=</span> <span class="fl">20</span> <span class="op">/</span> <span class="fl">100</span><span class="op">)</span></span>
<span></span>
<span><span class="va">pseq_df</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/phy_to_ldf.html">phy_to_ldf</a></span><span class="op">(</span><span class="va">ps0</span>, transform.counts <span class="op">=</span> <span class="cn">NULL</span><span class="op">)</span></span>
<span><span class="co">#&gt; An additonal column Sam_rep with sample names is created for reference purpose</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/knitr/man/kable.html" class="external-link">kable</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">pseq_df</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<table class="table">
<colgroup>
<col width="4%">
<col width="4%">
<col width="5%">
<col width="5%">
<col width="5%">
<col width="6%">
<col width="5%">
<col width="2%">
<col width="5%">
<col width="3%">
<col width="6%">
<col width="27%">
<col width="4%">
<col width="1%">
<col width="4%">
<col width="4%">
<col width="3%">
</colgroup>
<thead><tr class="header">
<th align="left">OTUID</th>
<th align="left">Domain</th>
<th align="left">Phylum</th>
<th align="left">Class</th>
<th align="left">Order</th>
<th align="left">Family</th>
<th align="left">Genus</th>
<th align="left">Species</th>
<th align="left">Sam_rep</th>
<th align="right">Abundance</th>
<th align="left">investigation_type</th>
<th align="left">project_name</th>
<th align="left">DiseaseState</th>
<th align="right">age</th>
<th align="left">body_product</th>
<th align="left">FOBT.result</th>
<th align="left">material</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left">d__denovo66</td>
<td align="left">k__Bacteria</td>
<td align="left">p__Firmicutes</td>
<td align="left">c__Clostridia</td>
<td align="left">o__Clostridiales</td>
<td align="left">f__Lachnospiraceae</td>
<td align="left">g__Ruminococcus2</td>
<td align="left">s__</td>
<td align="left">Adenoma10-2757</td>
<td align="right">27</td>
<td align="left">metagenomic</td>
<td align="left">The Gut Microbiome Improves Predictive Models for
Diagnosis of Colorectal Cancer</td>
<td align="left">nonCRC</td>
<td align="right">37</td>
<td align="left">feces</td>
<td align="left">negative</td>
<td align="left">feces</td>
</tr>
<tr class="even">
<td align="left">d__denovo175</td>
<td align="left">k__Bacteria</td>
<td align="left">p__Bacteroidetes</td>
<td align="left">c__Bacteroidia</td>
<td align="left">o__Bacteroidales</td>
<td align="left">f__Rikenellaceae</td>
<td align="left">g__Alistipes</td>
<td align="left">s__</td>
<td align="left">Adenoma10-2757</td>
<td align="right">0</td>
<td align="left">metagenomic</td>
<td align="left">The Gut Microbiome Improves Predictive Models for
Diagnosis of Colorectal Cancer</td>
<td align="left">nonCRC</td>
<td align="right">37</td>
<td align="left">feces</td>
<td align="left">negative</td>
<td align="left">feces</td>
</tr>
<tr class="odd">
<td align="left">d__denovo165</td>
<td align="left">k__Bacteria</td>
<td align="left">p__Firmicutes</td>
<td align="left">c__Clostridia</td>
<td align="left">o__Clostridiales</td>
<td align="left">f__Ruminococcaceae</td>
<td align="left">g__</td>
<td align="left">s__</td>
<td align="left">Adenoma10-2757</td>
<td align="right">49</td>
<td align="left">metagenomic</td>
<td align="left">The Gut Microbiome Improves Predictive Models for
Diagnosis of Colorectal Cancer</td>
<td align="left">nonCRC</td>
<td align="right">37</td>
<td align="left">feces</td>
<td align="left">negative</td>
<td align="left">feces</td>
</tr>
<tr class="even">
<td align="left">d__denovo167</td>
<td align="left">k__Bacteria</td>
<td align="left">p__Firmicutes</td>
<td align="left">c__Clostridia</td>
<td align="left">o__Clostridiales</td>
<td align="left">f__Lachnospiraceae</td>
<td align="left">g__Coprococcus</td>
<td align="left">s__</td>
<td align="left">Adenoma10-2757</td>
<td align="right">155</td>
<td align="left">metagenomic</td>
<td align="left">The Gut Microbiome Improves Predictive Models for
Diagnosis of Colorectal Cancer</td>
<td align="left">nonCRC</td>
<td align="right">37</td>
<td align="left">feces</td>
<td align="left">negative</td>
<td align="left">feces</td>
</tr>
<tr class="odd">
<td align="left">d__denovo166</td>
<td align="left">k__Bacteria</td>
<td align="left">p__Firmicutes</td>
<td align="left">c__Clostridia</td>
<td align="left">o__Clostridiales</td>
<td align="left">f__</td>
<td align="left">g__</td>
<td align="left">s__</td>
<td align="left">Adenoma10-2757</td>
<td align="right">0</td>
<td align="left">metagenomic</td>
<td align="left">The Gut Microbiome Improves Predictive Models for
Diagnosis of Colorectal Cancer</td>
<td align="left">nonCRC</td>
<td align="right">37</td>
<td align="left">feces</td>
<td align="left">negative</td>
<td align="left">feces</td>
</tr>
<tr class="even">
<td align="left">d__denovo161</td>
<td align="left">k__Bacteria</td>
<td align="left">p__Firmicutes</td>
<td align="left">c__Clostridia</td>
<td align="left">o__Clostridiales</td>
<td align="left">f__Lachnospiraceae</td>
<td align="left">g__Roseburia</td>
<td align="left">s__</td>
<td align="left">Adenoma10-2757</td>
<td align="right">11</td>
<td align="left">metagenomic</td>
<td align="left">The Gut Microbiome Improves Predictive Models for
Diagnosis of Colorectal Cancer</td>
<td align="left">nonCRC</td>
<td align="right">37</td>
<td align="left">feces</td>
<td align="left">negative</td>
<td align="left">feces</td>
</tr>
</tbody>
</table>
</div>
</div>
<div class="section level2">
<h2 id="taxa-overview">Taxa overview<a class="anchor" aria-label="anchor" href="#taxa-overview"></a>
</h2>
<p>One of the first questions arise regarding which taxa are present ,
how they are distributed in the data. This can be done with following
functionality.</p>
<div class="section level3">
<h3 id="check-distribution">Check distribution<a class="anchor" aria-label="anchor" href="#check-distribution"></a>
</h3>
<p>A quick check to see how different taxa are distributed in your
data.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">microbiomeutilities</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">RColorBrewer</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://patchwork.data-imaginist.com" class="external-link">patchwork</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://rpkgs.datanovia.com/ggpubr/" class="external-link">ggpubr</a></span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"zackular2014"</span><span class="op">)</span></span>
<span><span class="va">pseq</span> <span class="op">&lt;-</span> <span class="va">zackular2014</span></span>
<span></span>
<span><span class="co"># check healthy</span></span>
<span><span class="va">health_ps</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/phyloseq/man/subset_samples-methods.html" class="external-link">subset_samples</a></span><span class="op">(</span><span class="va">pseq</span>, <span class="va">DiseaseState</span><span class="op">==</span><span class="st">"H"</span><span class="op">)</span></span>
<span><span class="va">p_hc</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/taxa_distribution.html">taxa_distribution</a></span><span class="op">(</span><span class="va">health_ps</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="../reference/theme_biome_utils.html">theme_biome_utils</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>title <span class="op">=</span> <span class="st">"Healthy"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># check CRC</span></span>
<span><span class="va">crc_ps</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/phyloseq/man/subset_samples-methods.html" class="external-link">subset_samples</a></span><span class="op">(</span><span class="va">pseq</span>, <span class="va">DiseaseState</span><span class="op">==</span><span class="st">"CRC"</span><span class="op">)</span></span>
<span><span class="va">p_crc</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/taxa_distribution.html">taxa_distribution</a></span><span class="op">(</span><span class="va">crc_ps</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="../reference/theme_biome_utils.html">theme_biome_utils</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>title <span class="op">=</span> <span class="st">"CRC"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># harnessing the power of patchwork</span></span>
<span><span class="va">p_hc</span> <span class="op">/</span> <span class="va">p_crc</span> <span class="op">+</span> <span class="fu"><a href="https://patchwork.data-imaginist.com/reference/plot_layout.html" class="external-link">plot_layout</a></span><span class="op">(</span>guides <span class="op">=</span> <span class="st">"collect"</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://patchwork.data-imaginist.com/reference/plot_annotation.html" class="external-link">plot_annotation</a></span><span class="op">(</span>tag_levels <span class="op">=</span> <span class="st">"A"</span><span class="op">)</span></span></code></pre></div>
<p><img src="microbiomeutilities_files/figure-html/unnamed-chunk-9-1.png" width="768"></p>
<p>There are Cyanobacteria/Chloroplast related sequences which can be
removed if not expected in the samples.</p>
</div>
<div class="section level3">
<h3 id="dominant-taxa">Dominant taxa<a class="anchor" aria-label="anchor" href="#dominant-taxa"></a>
</h3>
<p>Sometimes, we are interested in identifying the most dominant taxa in
each sample. We may also wish to check what percent of samples within a
given group are these taxa dominating.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">microbiomeutilities</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://dplyr.tidyverse.org" class="external-link">dplyr</a></span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"zackular2014"</span><span class="op">)</span></span>
<span><span class="va">p0</span> <span class="op">&lt;-</span> <span class="va">zackular2014</span></span>
<span><span class="va">p0.gen</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/microbiome/man/aggregate_taxa.html" class="external-link">aggregate_taxa</a></span><span class="op">(</span><span class="va">p0</span>,<span class="st">"Genus"</span><span class="op">)</span></span>
<span><span class="va">x.d</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/dominant_taxa.html">dominant_taxa</a></span><span class="op">(</span><span class="va">p0</span>,level <span class="op">=</span> <span class="st">"Genus"</span>, group<span class="op">=</span><span class="st">"DiseaseState"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">x.d</span><span class="op">$</span><span class="va">dominant_overview</span><span class="op">)</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 6 × 5</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># Groups:   DiseaseState [3]</span></span></span>
<span><span class="co">#&gt;   DiseaseState dominant_taxa      n rel.freq rel.freq.pct</span></span>
<span><span class="co">#&gt;   <span style="color: #949494; font-style: italic;">&lt;fct&gt;</span>        <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>          <span style="color: #949494; font-style: italic;">&lt;int&gt;</span>    <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>       </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">1</span> H            g__Bacteroides    15     50   50%         </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">2</span> nonCRC       g__Blautia        13     46.4 46%         </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">3</span> CRC          g__Bacteroides     9     30   30%         </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">4</span> CRC          g__Blautia         9     30   30%         </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">5</span> H            g__Blautia         7     23.3 23%         </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">6</span> nonCRC       g__Bacteroides     6     21.4 21%</span></span></code></pre></div>
<p>As seen in the table above, 50% of the samples in <code>H</code>
group are dominated by *g__Bacteroides* and so on…</p>
</div>
<div class="section level3">
<h3 id="get-taxa-summary">Get taxa summary<a class="anchor" aria-label="anchor" href="#get-taxa-summary"></a>
</h3>
<p>This can be used for entire dataset.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">microbiomeutilities</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"zackular2014"</span><span class="op">)</span></span>
<span><span class="va">p0</span> <span class="op">&lt;-</span> <span class="va">zackular2014</span></span>
<span><span class="va">tx.sum1</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/taxa_summary.html">taxa_summary</a></span><span class="op">(</span><span class="va">p0</span>, <span class="st">"Phylum"</span><span class="op">)</span></span>
<span><span class="co">#&gt; Data provided is not compositional </span></span>
<span><span class="co">#&gt;  will first transform</span></span>
<span><span class="va">tx.sum1</span></span>
<span><span class="co">#&gt;                            Taxa          Max.Rel.Ab          Mean.Rel.Ab</span></span>
<span><span class="co">#&gt; 1              p__Euryarchaeota   0.218508035656191   0.0200267274400673</span></span>
<span><span class="co">#&gt; 2                           p__  0.0359956743395643  0.00213630393381328</span></span>
<span><span class="co">#&gt; 3             p__Actinobacteria   0.189937172231165   0.0294096453288836</span></span>
<span><span class="co">#&gt; 4              p__Bacteroidetes   0.703769740193581    0.244709411770076</span></span>
<span><span class="co">#&gt; 5  p__Cyanobacteria/Chloroplast 0.00391786546652736 0.000248991452140149</span></span>
<span><span class="co">#&gt; 6        p__Deinococcus-Thermus  0.0306807689079157  0.00136619388828292</span></span>
<span><span class="co">#&gt; 7                 p__Firmicutes   0.935779308307329    0.609116019352896</span></span>
<span><span class="co">#&gt; 8               p__Fusobacteria   0.390135934239562   0.0079458294240542</span></span>
<span><span class="co">#&gt; 9             p__Proteobacteria   0.403932082216265     0.04116346455827</span></span>
<span><span class="co">#&gt; 10             p__Synergistetes  0.0106576610162058 0.000207008030950549</span></span>
<span><span class="co">#&gt; 11           p__Verrucomicrobia   0.320077232101149   0.0436704048205656</span></span>
<span><span class="co">#&gt;           Median.Rel.Ab             Std.dev</span></span>
<span><span class="co">#&gt; 1  4.08341966447163e-05  0.0445500008069431</span></span>
<span><span class="co">#&gt; 2  0.000335981950200747 0.00577447214134393</span></span>
<span><span class="co">#&gt; 3    0.0152970058974793  0.0361457627428506</span></span>
<span><span class="co">#&gt; 4     0.193736718210234   0.183986562646175</span></span>
<span><span class="co">#&gt; 5  2.58996131628288e-05 0.00070384339925814</span></span>
<span><span class="co">#&gt; 6                     0 0.00449398785197828</span></span>
<span><span class="co">#&gt; 7     0.598709982286244   0.182320763133074</span></span>
<span><span class="co">#&gt; 8                     0  0.0514603201611758</span></span>
<span><span class="co">#&gt; 9    0.0130121759294526  0.0774502267516943</span></span>
<span><span class="co">#&gt; 10                    0 0.00119199461324549</span></span>
<span><span class="co">#&gt; 11  0.00665688163392832  0.0679829229763011</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="get-taxa-summary-by-groups">Get taxa summary by group(s)<a class="anchor" aria-label="anchor" href="#get-taxa-summary-by-groups"></a>
</h3>
<p>For group specific abundances of taxa
<code>get_group_abundances</code>.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">microbiomeutilities</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"zackular2014"</span><span class="op">)</span></span>
<span><span class="va">p0</span> <span class="op">&lt;-</span> <span class="va">zackular2014</span></span>
<span><span class="va">grp_abund</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/get_group_abundances.html">get_group_abundances</a></span><span class="op">(</span><span class="va">p0</span>, </span>
<span>                                  level <span class="op">=</span> <span class="st">"Phylum"</span>, </span>
<span>                                  group<span class="op">=</span><span class="st">"DiseaseState"</span>,</span>
<span>                                  transform <span class="op">=</span> <span class="st">"compositional"</span><span class="op">)</span></span></code></pre></div>
<p>How to visualize these data?</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">mycols</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"brown3"</span>, <span class="st">"steelblue"</span>,<span class="st">"grey50"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># clean names </span></span>
<span><span class="va">grp_abund</span><span class="op">$</span><span class="va">OTUID</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">gsub</a></span><span class="op">(</span><span class="st">"p__"</span>, <span class="st">""</span>,<span class="va">grp_abund</span><span class="op">$</span><span class="va">OTUID</span><span class="op">)</span></span>
<span><span class="va">grp_abund</span><span class="op">$</span><span class="va">OTUID</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/ifelse.html" class="external-link">ifelse</a></span><span class="op">(</span><span class="va">grp_abund</span><span class="op">$</span><span class="va">OTUID</span> <span class="op">==</span> <span class="st">""</span>, </span>
<span>                          <span class="st">"Unclassified"</span>, <span class="va">grp_abund</span><span class="op">$</span><span class="va">OTUID</span><span class="op">)</span></span>
<span></span>
<span></span>
<span><span class="va">mean.plot</span> <span class="op">&lt;-</span> <span class="va">grp_abund</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="co"># input data</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x<span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/stats/reorder.factor.html" class="external-link">reorder</a></span><span class="op">(</span><span class="va">OTUID</span>, <span class="va">mean_abundance</span><span class="op">)</span>, <span class="co"># reroder based on mean abundance</span></span>
<span>             y<span class="op">=</span> <span class="va">mean_abundance</span>,</span>
<span>             fill<span class="op">=</span><span class="va">DiseaseState</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> <span class="co"># x and y axis </span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_bar.html" class="external-link">geom_bar</a></span><span class="op">(</span>     stat <span class="op">=</span> <span class="st">"identity"</span>, </span>
<span>                position<span class="op">=</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/position_dodge.html" class="external-link">position_dodge</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html" class="external-link">scale_fill_manual</a></span><span class="op">(</span><span class="st">"DiseaseState"</span>, values<span class="op">=</span><span class="va">mycols</span><span class="op">)</span> <span class="op">+</span> <span class="co"># manually specify colors</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_bw</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> <span class="co"># add a widely used ggplot2 theme</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ylab</a></span><span class="op">(</span><span class="st">"Mean Relative Abundance"</span><span class="op">)</span> <span class="op">+</span> <span class="co"># label y axis</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">xlab</a></span><span class="op">(</span><span class="st">"Phylum"</span><span class="op">)</span> <span class="op">+</span> <span class="co"># label x axis</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/coord_flip.html" class="external-link">coord_flip</a></span><span class="op">(</span><span class="op">)</span> <span class="co"># rotate plot </span></span>
<span></span>
<span><span class="va">mean.plot</span></span></code></pre></div>
<p><img src="microbiomeutilities_files/figure-html/unnamed-chunk-13-1.png" width="576" style="display: block; margin: auto;"></p>
<p>This plot is diagnostic to have an idea about the taxonomy in each
group. Statements based on comparisons between groups may not make sense
here because only mean abundance values are plotted, and not standard
deviation within groups. For visualizing error bars and more check <a href="http://www.sthda.com/english/wiki/ggplot2-error-bars-quick-start-guide-r-software-and-data-visualization" class="external-link">Statistical
tools for high-throughput data analysis</a>.</p>
</div>
<div class="section level3">
<h3 id="find-samples-dominated-by-specific-taxa">Find samples dominated by specific taxa<a class="anchor" aria-label="anchor" href="#find-samples-dominated-by-specific-taxa"></a>
</h3>
<p>Finding samples dominated by user provided taxa in a phyloseq object.
This is useful especially if user suspects a taxa to be contaminant and
wishes to identify which samples are dominated by the contaminant
taxa.</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">microbiomeutilities</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"zackular2014"</span><span class="op">)</span></span>
<span><span class="va">p0</span> <span class="op">&lt;-</span> <span class="va">zackular2014</span></span>
<span><span class="va">p0.f</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/microbiome/man/aggregate_taxa.html" class="external-link">aggregate_taxa</a></span><span class="op">(</span><span class="va">p0</span>, <span class="st">"Genus"</span><span class="op">)</span></span>
<span><span class="va">bac_dom</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/find_samples_taxa.html">find_samples_taxa</a></span><span class="op">(</span><span class="va">p0.f</span>, taxa <span class="op">=</span> <span class="st">"g__Bacteroides"</span><span class="op">)</span></span>
<span><span class="va">bac_dom</span></span>
<span><span class="co">#&gt;  [1] "Adenoma14-2817"    "Adenoma20-2995"    "Adenoma21-3129"   </span></span>
<span><span class="co">#&gt;  [4] "Adenoma22-3133"    "Adenoma24-3217"    "Adenoma9-2717"    </span></span>
<span><span class="co">#&gt;  [7] "Cancer10-2567"     "Cancer11-2573"     "Cancer1-2355"     </span></span>
<span><span class="co">#&gt; [10] "Cancer17-2625"     "Cancer20-2667"     "Cancer22-2703"    </span></span>
<span><span class="co">#&gt; [13] "Cancer30-2863"     "Cancer5-2523"      "Cancer9-2547"     </span></span>
<span><span class="co">#&gt; [16] "Healthy10-3087"    "Healthy11-3123"    "Healthy1-2027"    </span></span>
<span><span class="co">#&gt; [19] "Healthy13-3241"    "Healthy14-3245"    "Healthy16-3251"   </span></span>
<span><span class="co">#&gt; [22] "Healthy17-3257"    "Healthy18-3293"    "Healthy19-3339"   </span></span>
<span><span class="co">#&gt; [25] "Healthy20-3341"    "Healthy26-3407"    "Healthy30-3291-66"</span></span>
<span><span class="co">#&gt; [28] "Healthy6-3061"     "Healthy8-3073"     "Healthy9-3075"</span></span>
<span></span>
<span><span class="co">#get samples dominated by g__Bacteroides</span></span>
<span><span class="va">ps.sub</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/phyloseq/man/prune_samples-methods.html" class="external-link">prune_samples</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/phyloseq/man/sample_names-methods.html" class="external-link">sample_names</a></span><span class="op">(</span><span class="va">p0.f</span><span class="op">)</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="va">bac_dom</span>, <span class="va">p0.f</span><span class="op">)</span></span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 id="alpha-diversity">Alpha diversity<a class="anchor" aria-label="anchor" href="#alpha-diversity"></a>
</h2>
<div class="section level3">
<h3 id="rarefaction-curves-for-alpha-diversity-indices">Rarefaction curves for alpha diversity indices<a class="anchor" aria-label="anchor" href="#rarefaction-curves-for-alpha-diversity-indices"></a>
</h3>
<p>A common approach to check for sequencing depth and diversity
measures.<br>
Here, we can access the numerous alpha diversity measures supported by
<a href="https://microbiome.github.io/tutorials/Alphadiversity.html" class="external-link">microbiome
R package</a>.</p>
<p><em>NOTE:</em> This can take sometime to complete.</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">microbiomeutilities</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"zackular2014"</span><span class="op">)</span></span>
<span><span class="va">p0</span> <span class="op">&lt;-</span> <span class="va">zackular2014</span></span>
<span><span class="co"># set seed</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">1</span><span class="op">)</span></span>
<span><span class="va">subsamples</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">5000</span>, by<span class="op">=</span><span class="fl">100</span><span class="op">)</span><span class="op">[</span><span class="op">-</span><span class="fl">1</span><span class="op">]</span></span>
<span><span class="co">#subsamples = c(10, 5000, 10000, 20000, 30000)</span></span>
<span></span>
<span><span class="va">p</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/plot_alpha_rcurve.html">plot_alpha_rcurve</a></span><span class="op">(</span><span class="va">p0</span>, index<span class="op">=</span><span class="st">"observed"</span>, </span>
<span>                       subsamples <span class="op">=</span> <span class="va">subsamples</span>,</span>
<span>                       lower.conf <span class="op">=</span> <span class="fl">0.025</span>, </span>
<span>                       upper.conf <span class="op">=</span> <span class="fl">0.975</span>,</span>
<span>                       group<span class="op">=</span><span class="st">"DiseaseState"</span>,</span>
<span>                       label.color <span class="op">=</span> <span class="st">"brown3"</span>,</span>
<span>                       label.size <span class="op">=</span> <span class="fl">3</span>,</span>
<span>                       label.min <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span> </span>
<span><span class="co"># change color of line </span></span>
<span><span class="va">mycols</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"brown3"</span>, <span class="st">"steelblue"</span>,<span class="st">"grey50"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">p</span> <span class="op">&lt;-</span> <span class="va">p</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html" class="external-link">scale_color_manual</a></span><span class="op">(</span>values <span class="op">=</span> <span class="va">mycols</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html" class="external-link">scale_fill_manual</a></span><span class="op">(</span>values <span class="op">=</span> <span class="va">mycols</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">p</span><span class="op">)</span></span></code></pre></div>
<p><img src="microbiomeutilities_files/figure-html/unnamed-chunk-15-1.png" width="576" style="display: block; margin: auto;"></p>
</div>
<div class="section level3">
<h3 id="plot-alpha-diversities">Plot alpha diversities<a class="anchor" aria-label="anchor" href="#plot-alpha-diversities"></a>
</h3>
<p>Utility plot function for diversity measures calculated by
<code>microbiome</code> package.</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="http://microbiome.github.io/microbiome" class="external-link">microbiome</a></span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"zackular2014"</span><span class="op">)</span></span>
<span><span class="va">ps0</span> <span class="op">&lt;-</span> <span class="va">zackular2014</span></span>
<span><span class="va">mycols</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"brown3"</span>, <span class="st">"steelblue"</span>,<span class="st">"grey50"</span><span class="op">)</span></span>
<span><span class="va">p</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/plot_alpha_diversities.html">plot_alpha_diversities</a></span><span class="op">(</span><span class="va">ps0</span>,</span>
<span>                            type <span class="op">=</span> <span class="st">"dominance"</span>,</span>
<span>                            index.val <span class="op">=</span> <span class="st">"all"</span>,</span>
<span>                            plot.type <span class="op">=</span> <span class="st">"stripchart"</span>,</span>
<span>                            variableA <span class="op">=</span> <span class="st">"DiseaseState"</span>,</span>
<span>                            palette <span class="op">=</span> <span class="va">mycols</span><span class="op">)</span></span>
<span></span>
<span><span class="va">p</span> <span class="op">&lt;-</span> <span class="va">p</span> <span class="op">+</span> <span class="fu"><a href="../reference/theme_biome_utils.html">theme_biome_utils</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>legend.position <span class="op">=</span> <span class="st">"top"</span>,</span>
<span>                 text <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_text</a></span><span class="op">(</span>size<span class="op">=</span><span class="fl">14</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">p</span><span class="op">)</span></span></code></pre></div>
<p><img src="microbiomeutilities_files/figure-html/unnamed-chunk-16-1.png" width="576" style="display: block; margin: auto;"></p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">comps</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/make_pairs.html">make_pairs</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/phyloseq/man/sample_data-methods.html" class="external-link">sample_data</a></span><span class="op">(</span><span class="va">ps0</span><span class="op">)</span><span class="op">$</span><span class="va">DiseaseState</span><span class="op">)</span></span>
<span></span>
<span><span class="va">p</span> <span class="op">&lt;-</span> <span class="va">p</span> <span class="op">+</span> <span class="fu"><a href="https://rpkgs.datanovia.com/ggpubr/reference/stat_compare_means.html" class="external-link">stat_compare_means</a></span><span class="op">(</span></span>
<span>      comparisons <span class="op">=</span> <span class="va">comps</span>,</span>
<span>      label <span class="op">=</span> <span class="st">"p.format"</span>,</span>
<span>      tip.length <span class="op">=</span> <span class="fl">0.05</span>,</span>
<span>      method <span class="op">=</span> <span class="st">"wilcox.test"</span><span class="op">)</span></span>
<span><span class="va">p</span></span>
<span><span class="co">#&gt; Warning in wilcox.test.default(c(7855, 4967, 16964, 6997, 11425, 17013, : cannot</span></span>
<span><span class="co">#&gt; compute exact p-value with ties</span></span></code></pre></div>
<p><img src="microbiomeutilities_files/figure-html/unnamed-chunk-17-1.png" width="768" style="display: block; margin: auto;"></p>
<p>Alternatively, one can plot one index at a time with pair-wise
stats.</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/erocoar/gghalves" class="external-link">gghalves</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">microbiomeutilities</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"zackular2014"</span><span class="op">)</span></span>
<span><span class="va">p0</span> <span class="op">&lt;-</span> <span class="va">zackular2014</span></span>
<span><span class="va">mycols</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"brown3"</span>, <span class="st">"steelblue"</span>,<span class="st">"grey50"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">p.m</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/plot_diversity_stats.html">plot_diversity_stats</a></span><span class="op">(</span><span class="va">p0</span>, group <span class="op">=</span> <span class="st">"DiseaseState"</span>, </span>
<span>                            index <span class="op">=</span> <span class="st">"diversity_shannon"</span>, </span>
<span>                            group.order <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"H"</span>, <span class="st">"CRC"</span>, <span class="st">"nonCRC"</span><span class="op">)</span>, </span>
<span>                            group.colors <span class="op">=</span> <span class="va">mycols</span>,</span>
<span>                            label.format<span class="op">=</span><span class="st">"p.format"</span>,</span>
<span>                            stats <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="co">#&gt; Warning: `guides(&lt;scale&gt; = FALSE)` is deprecated. Please use `guides(&lt;scale&gt; =</span></span>
<span><span class="co">#&gt; "none")` instead.</span></span>
<span><span class="va">p.m</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ylab</a></span><span class="op">(</span><span class="st">"Shannon Diversity"</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">xlab</a></span><span class="op">(</span><span class="st">""</span><span class="op">)</span></span></code></pre></div>
<p><img src="microbiomeutilities_files/figure-html/unnamed-chunk-18-1.png" width="384" style="display: block; margin: auto;"></p>
</div>
</div>
<div class="section level2">
<h2 id="composition-plots">Composition plots<a class="anchor" aria-label="anchor" href="#composition-plots"></a>
</h2>
<div class="section level3">
<h3 id="ternary-plot">Ternary plot<a class="anchor" aria-label="anchor" href="#ternary-plot"></a>
</h3>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="http://microbiome.github.io/microbiome" class="external-link">microbiome</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">microbiomeutilities</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://dplyr.tidyverse.org" class="external-link">dplyr</a></span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"zackular2014"</span><span class="op">)</span></span>
<span><span class="va">p0</span> <span class="op">&lt;-</span> <span class="va">zackular2014</span></span>
<span><span class="va">tern_df</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/prep_ternary.html">prep_ternary</a></span><span class="op">(</span><span class="va">p0</span>, group<span class="op">=</span><span class="st">"DiseaseState"</span>, </span>
<span>               abund.thres<span class="op">=</span><span class="fl">0.000001</span>, level<span class="op">=</span> <span class="st">"Genus"</span>, prev.thres<span class="op">=</span><span class="fl">0.01</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">tern_df</span><span class="op">)</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 6 × 11</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># Groups:   unique [6]</span></span></span>
<span><span class="co">#&gt;   unique         CRC      H  nonCRC Domain Phylum Class Order Family Genus OTUID</span></span>
<span><span class="co">#&gt;   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>        <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">1</span> Acetanaer… 0.009<span style="text-decoration: underline;">95</span> 0.013<span style="text-decoration: underline;">5</span> 0.008<span style="text-decoration: underline;">58</span> Bacte… Firmi… Clos… Clos… Rumin… Acet… Acet…</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">2</span> Acidamino… 0.115   0.164  0.084<span style="text-decoration: underline;">6</span>  Bacte… Firmi… Nega… Sele… Acida… Acid… Acid…</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">3</span> Acinetoba… 0.010<span style="text-decoration: underline;">8</span>  0.031<span style="text-decoration: underline;">2</span> 0.007<span style="text-decoration: underline;">62</span> Bacte… Prote… Gamm… Pseu… Morax… Acin… Acin…</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">4</span> Actinomyc… 0.016<span style="text-decoration: underline;">9</span>  0.033<span style="text-decoration: underline;">8</span> 0.021<span style="text-decoration: underline;">8</span>  Bacte… Actin… Acti… Acti… Actin… Acti… Acti…</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">5</span> Akkermans… 1.26    1.05   1.53    Bacte… Verru… Verr… Verr… Verru… Akke… Akke…</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">6</span> Alistipes  0.929   0.894  0.654   Bacte… Bacte… Bact… Bact… Riken… Alis… Alis…</span></span></code></pre></div>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="co"># install.packages("ggtern")</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">require</a></span><span class="op">(</span><span class="va"><a href="http://www.ggtern.com" class="external-link">ggtern</a></span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Replace empty with Other</span></span>
<span><span class="va">tern_df</span><span class="op">$</span><span class="va">Phylum</span><span class="op">[</span><span class="va">tern_df</span><span class="op">$</span><span class="va">Phylum</span><span class="op">==</span><span class="st">""</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="st">"Other"</span></span>
<span></span>
<span><span class="fu">ggtern</span><span class="op">(</span>data<span class="op">=</span><span class="va">tern_df</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x<span class="op">=</span><span class="va">CRC</span>, y<span class="op">=</span><span class="va">H</span>, z<span class="op">=</span><span class="va">nonCRC</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>color<span class="op">=</span> <span class="va">Phylum</span><span class="op">)</span>, </span>
<span>             alpha<span class="op">=</span><span class="fl">0.25</span>, </span>
<span>             show.legend<span class="op">=</span><span class="cn">T</span>, </span>
<span>             size<span class="op">=</span><span class="fl">3</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="co">#scale_size(range=c(0, 6)) + </span></span>
<span>  <span class="fu">geom_mask</span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_brewer.html" class="external-link">scale_colour_brewer</a></span><span class="op">(</span>palette <span class="op">=</span> <span class="st">"Paired"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="../reference/theme_biome_utils.html">theme_biome_utils</a></span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/detach.html" class="external-link">detach</a></span><span class="op">(</span><span class="st">"package:ggtern"</span>, unload<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="plot-taxa-boxplot">Plot taxa boxplot<a class="anchor" aria-label="anchor" href="#plot-taxa-boxplot"></a>
</h3>
<p>Plot relative abundance of top taxa specified by user.</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">microbiomeutilities</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">RColorBrewer</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"zackular2014"</span><span class="op">)</span></span>
<span><span class="va">ps0</span> <span class="op">&lt;-</span> <span class="va">zackular2014</span></span>
<span><span class="va">mycols</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"brown3"</span>, <span class="st">"steelblue"</span>, <span class="st">"grey50"</span><span class="op">)</span></span>
<span><span class="va">pn</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/plot_taxa_boxplot.html">plot_taxa_boxplot</a></span><span class="op">(</span><span class="va">ps0</span>,</span>
<span>                        taxonomic.level <span class="op">=</span> <span class="st">"Family"</span>,</span>
<span>                        top.otu <span class="op">=</span> <span class="fl">3</span>, </span>
<span>                        group <span class="op">=</span> <span class="st">"DiseaseState"</span>,</span>
<span>                        add.violin<span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>                        title <span class="op">=</span> <span class="st">"Top three family"</span>, </span>
<span>                        keep.other <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>                        group.order <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"H"</span>,<span class="st">"CRC"</span>,<span class="st">"nonCRC"</span><span class="op">)</span>,</span>
<span>                        group.colors <span class="op">=</span> <span class="va">mycols</span>,</span>
<span>                        dot.size <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">pn</span> <span class="op">+</span> <span class="fu"><a href="../reference/theme_biome_utils.html">theme_biome_utils</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="microbiomeutilities_files/figure-html/unnamed-chunk-21-1.png" width="672" style="display: block; margin: auto;"></p>
</div>
<div class="section level3">
<h3 id="plotting-selected-taxa">Plotting selected taxa<a class="anchor" aria-label="anchor" href="#plotting-selected-taxa"></a>
</h3>
<p>Using a list of taxa specified by the user for comparisons.</p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="http://microbiome.github.io/microbiome" class="external-link">microbiome</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">microbiomeutilities</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://rpkgs.datanovia.com/ggpubr/" class="external-link">ggpubr</a></span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"zackular2014"</span><span class="op">)</span></span>
<span><span class="va">p0</span> <span class="op">&lt;-</span> <span class="va">zackular2014</span></span>
<span><span class="va">p0.f</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/format_to_besthit.html">format_to_besthit</a></span><span class="op">(</span><span class="va">p0</span><span class="op">)</span></span>
<span><span class="co">#top_taxa(p0.f, 5)</span></span>
<span><span class="va">select.taxa</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"d__denovo1:g__Blautia"</span>, <span class="st">"d__denovo3:g__Bacteroides"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">mycols</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"brown3"</span>, <span class="st">"steelblue"</span>, <span class="st">"grey50"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">p</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/plot_listed_taxa.html">plot_listed_taxa</a></span><span class="op">(</span><span class="va">p0.f</span>, <span class="va">select.taxa</span>, </span>
<span>                      group<span class="op">=</span> <span class="st">"DiseaseState"</span>,</span>
<span>                      group.order <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"H"</span>,<span class="st">"CRC"</span>,<span class="st">"nonCRC"</span><span class="op">)</span>,</span>
<span>                      group.colors <span class="op">=</span> <span class="va">mycols</span>,</span>
<span>                      add.violin <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>                      violin.opacity <span class="op">=</span> <span class="fl">0.3</span>,</span>
<span>                      dot.opacity <span class="op">=</span> <span class="fl">0.25</span>,</span>
<span>                      box.opacity <span class="op">=</span> <span class="fl">0.25</span>,</span>
<span>                      panel.arrange<span class="op">=</span> <span class="st">"grid"</span><span class="op">)</span></span>
<span><span class="co">#&gt; An additonal column Sam_rep with sample names is created for reference purpose</span></span>
<span><span class="co">#&gt; Warning: `guides(&lt;scale&gt; = FALSE)` is deprecated. Please use `guides(&lt;scale&gt; =</span></span>
<span><span class="co">#&gt; "none")` instead.</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">p</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ylab</a></span><span class="op">(</span><span class="st">"Relative abundance"</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html" class="external-link">scale_y_continuous</a></span><span class="op">(</span>labels <span class="op">=</span> <span class="fu">scales</span><span class="fu">::</span><span class="va"><a href="https://scales.r-lib.org/reference/percent_format.html" class="external-link">percent</a></span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="microbiomeutilities_files/figure-html/unnamed-chunk-22-1.png" width="480" style="display: block; margin: auto;"></p>
<p>Adding statistical test with
<code><a href="https://rpkgs.datanovia.com/ggpubr/reference/stat_compare_means.html" class="external-link">ggpubr::stat_compare_means()</a></code></p>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="co"># If more than two variables</span></span>
<span><span class="va">comps</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/make_pairs.html">make_pairs</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/phyloseq/man/sample_data-methods.html" class="external-link">sample_data</a></span><span class="op">(</span><span class="va">p0.f</span><span class="op">)</span><span class="op">$</span><span class="va">DiseaseState</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">comps</span><span class="op">)</span></span>
<span><span class="co">#&gt; [[1]]</span></span>
<span><span class="co">#&gt; [1] "CRC" "H"  </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; [[2]]</span></span>
<span><span class="co">#&gt; [1] "CRC"    "nonCRC"</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; [[3]]</span></span>
<span><span class="co">#&gt; [1] "H"      "nonCRC"</span></span>
<span><span class="va">p</span> <span class="op">&lt;-</span> <span class="va">p</span> <span class="op">+</span> <span class="fu"><a href="https://rpkgs.datanovia.com/ggpubr/reference/stat_compare_means.html" class="external-link">stat_compare_means</a></span><span class="op">(</span></span>
<span>      comparisons <span class="op">=</span> <span class="va">comps</span>,</span>
<span>      label <span class="op">=</span> <span class="st">"p.format"</span>,</span>
<span>      tip.length <span class="op">=</span> <span class="fl">0.05</span>,</span>
<span>      method <span class="op">=</span> <span class="st">"wilcox.test"</span><span class="op">)</span></span>
<span><span class="va">p</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html" class="external-link">scale_y_continuous</a></span><span class="op">(</span>labels <span class="op">=</span> <span class="fu">scales</span><span class="fu">::</span><span class="va"><a href="https://scales.r-lib.org/reference/percent_format.html" class="external-link">percent</a></span><span class="op">)</span></span>
<span><span class="co">#&gt; Warning in wilcox.test.default(c(0.0737753988097334, 0.0369437665507036, :</span></span>
<span><span class="co">#&gt; cannot compute exact p-value with ties</span></span>
<span><span class="co">#&gt; Warning in wilcox.test.default(c(0.0218880793675399, 0.000317486610347303, :</span></span>
<span><span class="co">#&gt; cannot compute exact p-value with ties</span></span></code></pre></div>
<p><img src="microbiomeutilities_files/figure-html/unnamed-chunk-23-1.png" width="700" style="display: block; margin: auto;"></p>
<p>Plot top four Genera</p>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="http://microbiome.github.io/microbiome" class="external-link">microbiome</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">microbiomeutilities</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"zackular2014"</span><span class="op">)</span></span>
<span><span class="va">p0</span> <span class="op">&lt;-</span> <span class="va">zackular2014</span></span>
<span><span class="va">p0.f</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/microbiome/man/aggregate_taxa.html" class="external-link">aggregate_taxa</a></span><span class="op">(</span><span class="va">p0</span>, <span class="st">"Genus"</span><span class="op">)</span></span>
<span><span class="va">top_four</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/microbiome/man/top_taxa.html" class="external-link">top_taxa</a></span><span class="op">(</span><span class="va">p0.f</span>, <span class="fl">4</span><span class="op">)</span></span>
<span><span class="va">top_four</span></span>
<span><span class="co">#&gt; [1] "g__Bacteroides"      "g__Blautia"          "g__Faecalibacterium"</span></span>
<span><span class="co">#&gt; [4] "g__Akkermansia"</span></span>
<span></span>
<span><span class="va">mycols</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"brown3"</span>, <span class="st">"steelblue"</span>, <span class="st">"grey50"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">p</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/plot_listed_taxa.html">plot_listed_taxa</a></span><span class="op">(</span><span class="va">p0.f</span>, <span class="va">top_four</span>, </span>
<span>                      group<span class="op">=</span> <span class="st">"DiseaseState"</span>,</span>
<span>                      group.order <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"H"</span>,<span class="st">"CRC"</span>,<span class="st">"nonCRC"</span><span class="op">)</span>,</span>
<span>                      group.colors <span class="op">=</span> <span class="va">mycols</span>,</span>
<span>                      add.violin <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>                      violin.opacity <span class="op">=</span> <span class="fl">0.3</span>,</span>
<span>                      dot.opacity <span class="op">=</span> <span class="fl">0.25</span>,</span>
<span>                      box.opacity <span class="op">=</span> <span class="fl">0.25</span>,</span>
<span>                      panel.arrange<span class="op">=</span> <span class="st">"wrap"</span><span class="op">)</span></span>
<span><span class="co">#&gt; Warning: `guides(&lt;scale&gt; = FALSE)` is deprecated. Please use `guides(&lt;scale&gt; =</span></span>
<span><span class="co">#&gt; "none")` instead.</span></span>
<span></span>
<span><span class="va">comps</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/make_pairs.html">make_pairs</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/phyloseq/man/sample_data-methods.html" class="external-link">sample_data</a></span><span class="op">(</span><span class="va">p0.f</span><span class="op">)</span><span class="op">$</span><span class="va">DiseaseState</span><span class="op">)</span></span>
<span><span class="va">p</span> <span class="op">&lt;-</span> <span class="va">p</span> <span class="op">+</span> <span class="fu"><a href="https://rpkgs.datanovia.com/ggpubr/reference/stat_compare_means.html" class="external-link">stat_compare_means</a></span><span class="op">(</span></span>
<span>      comparisons <span class="op">=</span> <span class="va">comps</span>,</span>
<span>      label <span class="op">=</span> <span class="st">"p.format"</span>,</span>
<span>      tip.length <span class="op">=</span> <span class="fl">0.05</span>,</span>
<span>      method <span class="op">=</span> <span class="st">"wilcox.test"</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">p</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">ylab</a></span><span class="op">(</span><span class="st">"Relative abundance"</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html" class="external-link">scale_y_continuous</a></span><span class="op">(</span>labels <span class="op">=</span> <span class="fu">scales</span><span class="fu">::</span><span class="va"><a href="https://scales.r-lib.org/reference/percent_format.html" class="external-link">percent</a></span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; Warning in wilcox.test.default(c(0.0242631263865901, 5.19237758969832e-05, :</span></span>
<span><span class="co">#&gt; cannot compute exact p-value with ties</span></span></code></pre></div>
<p><img src="microbiomeutilities_files/figure-html/unnamed-chunk-24-1.png" width="576" style="display: block; margin: auto;"></p>
</div>
<div class="section level3">
<h3 id="abundance-prevalence-relationship">Abundance-Prevalence relationship<a class="anchor" aria-label="anchor" href="#abundance-prevalence-relationship"></a>
</h3>
<p>Plots mean Abundance-Prevalence for taxa. Mean abundance, mean
prevalence, and upper and lower confidence interval for each taxa is
calculated by random sub-sampling.<br>
This can be useful in identifying highly prevalent taxa and their mean
relative abundance in a group of samples. Taxa that are highly prevalent
with low variation in lower and upper CI can be identified at varying
values of mean relative abundance. These are likely core taxa in the
groups of samples.</p>
<p>See <a href="https://microbiome.github.io/tutorials/Core.html" class="external-link">core
microbiota analysis in microbiome R package</a></p>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">microbiomeutilities</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://dplyr.tidyverse.org" class="external-link">dplyr</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/slowkow/ggrepel" class="external-link">ggrepel</a></span><span class="op">)</span></span>
<span><span class="va">asv_ps</span> <span class="op">&lt;-</span> <span class="va">zackular2014</span></span>
<span><span class="va">asv_ps</span> <span class="op">&lt;-</span> <span class="fu">microbiome</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/microbiome/man/transform.html" class="external-link">transform</a></span><span class="op">(</span><span class="va">asv_ps</span>, <span class="st">"compositional"</span><span class="op">)</span></span>
<span><span class="co"># Select healthy samples</span></span>
<span><span class="va">asv_ps</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/phyloseq/man/subset_samples-methods.html" class="external-link">subset_samples</a></span><span class="op">(</span><span class="va">asv_ps</span>, <span class="va">DiseaseState</span><span class="op">==</span><span class="st">"H"</span><span class="op">)</span></span>
<span><span class="va">asv_ps</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/microbiome/man/core.html" class="external-link">core</a></span><span class="op">(</span><span class="va">asv_ps</span>,detection <span class="op">=</span> <span class="fl">0.0001</span>, prevalence <span class="op">=</span> <span class="fl">0.50</span><span class="op">)</span> <span class="co"># reduce size for example</span></span>
<span><span class="va">asv_ps</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/format_to_besthit.html">format_to_besthit</a></span><span class="op">(</span><span class="va">asv_ps</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">14285</span><span class="op">)</span></span>
<span><span class="va">p_v</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/plot_abund_prev.html">plot_abund_prev</a></span><span class="op">(</span><span class="va">asv_ps</span>, </span>
<span>                       label.core <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>                       color <span class="op">=</span> <span class="st">"Phylum"</span>, <span class="co"># NA or "blue"</span></span>
<span>                       mean.abund.thres <span class="op">=</span> <span class="fl">0.01</span>,</span>
<span>                       mean.prev.thres <span class="op">=</span> <span class="fl">0.99</span>,</span>
<span>                       dot.opacity <span class="op">=</span> <span class="fl">0.7</span>,</span>
<span>                       label.size <span class="op">=</span> <span class="fl">3</span>,</span>
<span>                       label.opacity <span class="op">=</span> <span class="fl">1.0</span>,</span>
<span>                       nudge.label<span class="op">=</span><span class="op">-</span><span class="fl">0.15</span>,</span>
<span>                       bs.iter<span class="op">=</span><span class="fl">9</span>, <span class="co"># increase for actual analysis e.g. 999</span></span>
<span>                       size <span class="op">=</span> <span class="fl">20</span>, <span class="co"># increase to match your nsamples(asv_ps)</span></span>
<span>                       replace <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>                       label.color<span class="op">=</span><span class="st">"#5f0f40"</span><span class="op">)</span>  </span>
<span><span class="va">p_v</span> <span class="op">&lt;-</span> <span class="va">p_v</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_abline.html" class="external-link">geom_vline</a></span><span class="op">(</span>xintercept <span class="op">=</span> <span class="fl">0.95</span>, lty<span class="op">=</span><span class="st">"dashed"</span>, alpha<span class="op">=</span><span class="fl">0.7</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_abline.html" class="external-link">geom_hline</a></span><span class="op">(</span>yintercept <span class="op">=</span> <span class="fl">0.01</span>,lty<span class="op">=</span><span class="st">"dashed"</span>, alpha<span class="op">=</span><span class="fl">0.7</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_brewer.html" class="external-link">scale_color_brewer</a></span><span class="op">(</span>palette <span class="op">=</span> <span class="st">"Dark2"</span><span class="op">)</span></span>
<span><span class="va">p_v</span></span></code></pre></div>
<p><img src="microbiomeutilities_files/figure-html/unnamed-chunk-25-1.png" width="576" style="display: block; margin: auto;"></p>
</div>
</div>
<div class="section level2">
<h2 id="heatmaps">Heatmaps<a class="anchor" aria-label="anchor" href="#heatmaps"></a>
</h2>
<div class="section level3">
<h3 id="heatmap-using-phyloseq-and-pheatmap">Heatmap using phyloseq and pheatmap<a class="anchor" aria-label="anchor" href="#heatmap-using-phyloseq-and-pheatmap"></a>
</h3>
<p>Useful for visualizing differences in top OTUs between sample
groups.</p>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">microbiomeutilities</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">pheatmap</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">RColorBrewer</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"zackular2014"</span><span class="op">)</span></span>
<span><span class="va">ps0</span> <span class="op">&lt;-</span> <span class="va">zackular2014</span></span>
<span></span>
<span><span class="co">#optional step if required to rename taxa_names</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/phyloseq/man/taxa_names-methods.html" class="external-link">taxa_names</a></span><span class="op">(</span><span class="va">ps0</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">gsub</a></span><span class="op">(</span><span class="st">"d__denovo"</span>, <span class="st">"OTU:"</span>, <span class="fu"><a href="https://rdrr.io/pkg/phyloseq/man/taxa_names-methods.html" class="external-link">taxa_names</a></span><span class="op">(</span><span class="va">ps0</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># create a gradient color palette for abundance</span></span>
<span><span class="va">grad_ab</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/grDevices/colorRamp.html" class="external-link">colorRampPalette</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"#faf3dd"</span>,<span class="st">"#f7d486"</span> ,<span class="st">"#5e6472"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">grad_ab_pal</span> <span class="op">&lt;-</span> <span class="fu">grad_ab</span><span class="op">(</span><span class="fl">10</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># create a color palette for varaibles of interest</span></span>
<span><span class="va">meta_colors</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"positive"</span> <span class="op">=</span> <span class="st">"#FFC857"</span>, </span>
<span>                      <span class="st">"negative"</span> <span class="op">=</span> <span class="st">"#05B083"</span><span class="op">)</span>, </span>
<span>                    <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"CRC"</span> <span class="op">=</span> <span class="st">"steelblue"</span>, </span>
<span>                      <span class="st">"nonCRC"</span> <span class="op">=</span> <span class="st">"grey50"</span>, </span>
<span>                      <span class="st">"H"</span><span class="op">=</span><span class="st">"brown3"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co"># add labels for pheatmap to detect</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">meta_colors</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"FOBT.result"</span>, <span class="st">"DiseaseState"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">p</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/plot_taxa_heatmap.html">plot_taxa_heatmap</a></span><span class="op">(</span><span class="va">ps0</span>,</span>
<span>                       subset.top <span class="op">=</span> <span class="fl">25</span>,</span>
<span>                       VariableA <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"DiseaseState"</span>,<span class="st">"FOBT.result"</span><span class="op">)</span>,</span>
<span>                       heatcolors <span class="op">=</span> <span class="va">grad_ab_pal</span>, <span class="co">#rev(brewer.pal(6, "RdPu")),</span></span>
<span>                       transformation <span class="op">=</span> <span class="st">"log10"</span>,</span>
<span>                       cluster_rows <span class="op">=</span> <span class="cn">T</span>,</span>
<span>                       cluster_cols <span class="op">=</span> <span class="cn">F</span>,</span>
<span>                       show_colnames <span class="op">=</span> <span class="cn">F</span>,</span>
<span>                       annotation_colors<span class="op">=</span><span class="va">meta_colors</span><span class="op">)</span></span>
<span><span class="co">#&gt; Top 25 OTUs selected</span></span>
<span><span class="co">#&gt; log10, if zeros in data then log10(1+x) will be used</span></span>
<span><span class="co">#&gt; First top taxa were selected and </span></span>
<span><span class="co">#&gt; then abundances tranformed to log10(1+X)</span></span>
<span><span class="co">#&gt; Warning in transform(phyobj1, "log10"): OTU table contains zeroes. Using log10(1</span></span>
<span><span class="co">#&gt; + x) transform.</span></span></code></pre></div>
<p><img src="microbiomeutilities_files/figure-html/unnamed-chunk-26-1.png" width="768" style="display: block; margin: auto;"></p>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#the plot is stored here</span></span>
<span><span class="va">p</span><span class="op">$</span><span class="va">plot</span></span>
<span></span>
<span><span class="co"># table used for plot is here</span></span>
<span><span class="va">p</span><span class="op">$</span><span class="va">tax_tab</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">3</span>,<span class="fl">1</span><span class="op">:</span><span class="fl">3</span><span class="op">]</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="elegant-option-with-ggplot2">Elegant option with ggplot2<a class="anchor" aria-label="anchor" href="#elegant-option-with-ggplot2"></a>
</h3>
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">microbiomeutilities</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"zackular2014"</span><span class="op">)</span></span>
<span><span class="va">p0</span> <span class="op">&lt;-</span> <span class="va">zackular2014</span></span>
<span><span class="va">p0.rel</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/microbiome/man/transform.html" class="external-link">transform</a></span><span class="op">(</span><span class="va">p0</span>, <span class="st">"compositional"</span><span class="op">)</span></span>
<span><span class="co">#heat.cols &lt;- c("#a8dadc","#457b9d", "#1d3557")</span></span>
<span><span class="co"># create a gradient color palette for abundance</span></span>
<span><span class="va">grad_ab</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/grDevices/colorRamp.html" class="external-link">colorRampPalette</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"#96d4ca"</span>,<span class="st">"#d3f3f1"</span>, <span class="st">"#7c65a9"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">heat.cols</span> <span class="op">&lt;-</span> <span class="fu">grad_ab</span><span class="op">(</span><span class="fl">10</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/simple_heatmap.html">simple_heatmap</a></span><span class="op">(</span><span class="va">p0.rel</span>,</span>
<span>               group.facet <span class="op">=</span> <span class="st">"DiseaseState"</span>,</span>
<span>               group.order <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>               abund.thres <span class="op">=</span> <span class="fl">0.01</span>,</span>
<span>               prev.thres <span class="op">=</span> <span class="fl">0.1</span>,</span>
<span>               level <span class="op">=</span> <span class="st">"Genus"</span>,</span>
<span>               scale.color <span class="op">=</span> <span class="st">"log10"</span>,</span>
<span>               na.fill <span class="op">=</span> <span class="st">"white"</span>,</span>
<span>               color.fill <span class="op">=</span> <span class="va">heat.cols</span>,</span>
<span>               taxa.arrange<span class="op">=</span><span class="cn">TRUE</span>,</span>
<span>               remove.other<span class="op">=</span><span class="cn">TRUE</span>,</span>
<span>               panel.arrange<span class="op">=</span><span class="st">"grid"</span>,</span>
<span>               ncol<span class="op">=</span><span class="cn">NULL</span>,</span>
<span>               nrow<span class="op">=</span><span class="cn">NULL</span><span class="op">)</span></span>
<span><span class="co">#&gt; Warning: Transformation introduced infinite values in discrete y-axis</span></span></code></pre></div>
<p><img src="microbiomeutilities_files/figure-html/unnamed-chunk-28-1.png" width="960" style="display: block; margin: auto;"></p>
<p>For heatmap options see <a href="https://microbiome.github.io/tutorials/Heatmap.html" class="external-link">microbiome::heat()</a>
and <a href="https://madsalbertsen.github.io/ampvis2/reference/amp_heatmap.html" class="external-link">ampvis2</a></p>
<p>For other longitudinal related functionality check the [Longitudinal
data analysis and visualization] section in the <em>Articles</em>
section.</p>
</div>
</div>
<div class="section level2">
<h2 id="microbiomehd-datasets-as-phyloseq-objects">MicrobiomeHD datasets as phyloseq objects<a class="anchor" aria-label="anchor" href="#microbiomehd-datasets-as-phyloseq-objects"></a>
</h2>
<p>We provide access to a subset of studies included in the
<code>MicrobiomeHD</code> database from Duvallet et al 2017: <a href="https://www.nature.com/articles/s41467-017-01973-8#ref-CR33" class="external-link">Meta-analysis
of gut microbiome studies identifies disease-specific and shared
responses</a>. Nature communications.</p>
<p>The phyloseq objects are stored and accessed from <a href="https://github.com/microsud/microbiomedatarepo" class="external-link">microbiomedatarepo</a>.</p>
<div class="sourceCode" id="cb29"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="va">study</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/list_microbiome_data.html">list_microbiome_data</a></span><span class="op">(</span>printtab <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">knitr</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/knitr/man/kable.html" class="external-link">kable</a></span><span class="op">(</span><span class="va">study</span><span class="op">)</span></span></code></pre></div>
<p>Below is the per study reference.</p>
<p><strong>NOTE</strong>: When using these studies, please cite <a href="https://www.nature.com/articles/s41467-017-01973-8#ref-CR33" class="external-link">Duvallet
et al. 2017</a> and the respective studies.</p>
<div class="sourceCode" id="cb30"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="va">file</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"microbiomeHD_ref.txt"</span>, package <span class="op">=</span> <span class="st">"microbiomeutilities"</span><span class="op">)</span></span>
<span><span class="va">reference</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html" class="external-link">read.table</a></span><span class="op">(</span><span class="va">file</span>, header <span class="op">=</span> <span class="cn">T</span>, sep <span class="op">=</span> <span class="st">"\t"</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">knitr</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/knitr/man/kable.html" class="external-link">kable</a></span><span class="op">(</span><span class="va">reference</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="experimental">Experimental<a class="anchor" aria-label="anchor" href="#experimental"></a>
</h2>
<div class="section level3">
<h3 id="plot-ordination-and-core">Plot ordination and core<a class="anchor" aria-label="anchor" href="#plot-ordination-and-core"></a>
</h3>
<div class="sourceCode" id="cb31"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">microbiomeutilities</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">RColorBrewer</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"zackular2014"</span><span class="op">)</span></span>
<span><span class="va">p0</span> <span class="op">&lt;-</span> <span class="va">zackular2014</span></span>
<span></span>
<span><span class="va">ps1</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/format_to_besthit.html">format_to_besthit</a></span><span class="op">(</span><span class="va">p0</span><span class="op">)</span></span>
<span></span>
<span><span class="co">#ps1 &lt;- subset_samples(ps1, DiseaseState == "H")</span></span>
<span><span class="va">ps1</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/phyloseq/man/prune_taxa-methods.html" class="external-link">prune_taxa</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/phyloseq/man/taxa_sums.html" class="external-link">taxa_sums</a></span><span class="op">(</span><span class="va">ps1</span><span class="op">)</span> <span class="op">&gt;</span> <span class="fl">0</span>, <span class="va">ps1</span><span class="op">)</span></span>
<span><span class="va">prev.thres</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq</a></span><span class="op">(</span><span class="fl">.05</span>, <span class="fl">1</span>, <span class="fl">.05</span><span class="op">)</span></span>
<span><span class="va">det.thres</span> <span class="op">&lt;-</span> <span class="fl">10</span><span class="op">^</span><span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">log10</a></span><span class="op">(</span><span class="fl">1e-4</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">log10</a></span><span class="op">(</span><span class="fl">.2</span><span class="op">)</span>, length <span class="op">=</span> <span class="fl">10</span><span class="op">)</span></span>
<span><span class="va">pseq.rel</span> <span class="op">&lt;-</span> <span class="fu">microbiome</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/microbiome/man/transform.html" class="external-link">transform</a></span><span class="op">(</span><span class="va">ps1</span>, <span class="st">"compositional"</span><span class="op">)</span></span>
<span><span class="co"># reduce size for example</span></span>
<span><span class="va">pseq.rel</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/microbiome/man/core.html" class="external-link">core</a></span><span class="op">(</span><span class="va">pseq.rel</span>, detection <span class="op">=</span> <span class="fl">0.001</span>, prevalence <span class="op">=</span> <span class="fl">20</span> <span class="op">/</span> <span class="fl">100</span><span class="op">)</span></span>
<span></span>
<span><span class="va">ord.bray</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/phyloseq/man/ordinate.html" class="external-link">ordinate</a></span><span class="op">(</span><span class="va">pseq.rel</span>, <span class="st">"NMDS"</span>, <span class="st">"bray"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">p</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/plot_ordiplot_core.html">plot_ordiplot_core</a></span><span class="op">(</span><span class="va">pseq.rel</span>, <span class="va">ord.bray</span>,</span>
<span>                        <span class="va">prev.thres</span>, <span class="va">det.thres</span>,</span>
<span>                        min.prevalence <span class="op">=</span> <span class="fl">0.9</span>,</span>
<span>                        color.opt <span class="op">=</span> <span class="st">"DiseaseState"</span>, </span>
<span>                        shape <span class="op">=</span> <span class="cn">NULL</span>, Sample <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">p</span><span class="op">)</span></span></code></pre></div>
<p><strong>Useful resources:</strong></p>
<p>The <code>microbiomeutilties</code> depends on the phyloseq data
structure and core <a href="http://joey711.github.io/phyloseq/" class="external-link">Phyloseq</a> functions.</p>
<p>Tools for microbiome analysis in R. <strong>Microbiome
package</strong> URL: <a href="http://microbiome.github.com/microbiome" class="external-link">microbiome
package</a>.</p>
<p>For more tutorials and examples of data analysis in R please
check:</p>
<ul>
<li>
<a href="https://goo.gl/CPChhd" class="external-link">Microbiome data analysis
SpringSchool2018</a><br>
</li>
<li><a href="http://microbiome.github.io/microbiome/" class="external-link">Microbiome R
package tutorials</a></li>
</ul>
<p><strong>Contributions are welcome:</strong></p>
<p><a href="https://github.com/microsud/microbiomeutilities/issues" class="external-link">Issue
Tracker</a><br><a href="https://github.com/microsud/microbiomeutilities/pulls" class="external-link">Pull
requests</a><br><a href="https://github.com/microsud/microbiomeutilities" class="external-link">Star us on the
Github page</a></p>
<div class="sourceCode" id="cb32"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; R version 4.2.1 (2022-06-23 ucrt)</span></span>
<span><span class="co">#&gt; Platform: x86_64-w64-mingw32/x64 (64-bit)</span></span>
<span><span class="co">#&gt; Running under: Windows 10 x64 (build 19044)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Matrix products: default</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; locale:</span></span>
<span><span class="co">#&gt; [1] LC_COLLATE=English_United States.utf8 </span></span>
<span><span class="co">#&gt; [2] LC_CTYPE=English_United States.utf8   </span></span>
<span><span class="co">#&gt; [3] LC_MONETARY=English_United States.utf8</span></span>
<span><span class="co">#&gt; [4] LC_NUMERIC=C                          </span></span>
<span><span class="co">#&gt; [5] LC_TIME=English_United States.utf8    </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; attached base packages:</span></span>
<span><span class="co">#&gt; [1] stats     graphics  grDevices utils     datasets  methods   base     </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; other attached packages:</span></span>
<span><span class="co">#&gt;  [1] pheatmap_1.0.12             ggrepel_0.9.1              </span></span>
<span><span class="co">#&gt;  [3] gghalves_0.1.3              ggpubr_0.4.0               </span></span>
<span><span class="co">#&gt;  [5] patchwork_1.1.2             RColorBrewer_1.1-3         </span></span>
<span><span class="co">#&gt;  [7] tibble_3.1.7                knitr_1.40                 </span></span>
<span><span class="co">#&gt;  [9] microbiomeutilities_1.00.16 dplyr_1.0.9                </span></span>
<span><span class="co">#&gt; [11] microbiome_1.18.0           ggplot2_3.3.6              </span></span>
<span><span class="co">#&gt; [13] phyloseq_1.40.0            </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; loaded via a namespace (and not attached):</span></span>
<span><span class="co">#&gt;  [1] nlme_3.1-157           bitops_1.0-7           fs_1.5.2              </span></span>
<span><span class="co">#&gt;  [4] rprojroot_2.0.3        GenomeInfoDb_1.32.2    backports_1.4.1       </span></span>
<span><span class="co">#&gt;  [7] tools_4.2.1            bslib_0.4.0            utf8_1.2.2            </span></span>
<span><span class="co">#&gt; [10] R6_2.5.1               vegan_2.6-2            DBI_1.1.3             </span></span>
<span><span class="co">#&gt; [13] BiocGenerics_0.42.0    mgcv_1.8-40            colorspace_2.0-3      </span></span>
<span><span class="co">#&gt; [16] permute_0.9-7          rhdf5filters_1.8.0     ade4_1.7-19           </span></span>
<span><span class="co">#&gt; [19] withr_2.5.0            tidyselect_1.1.2       compiler_4.2.1        </span></span>
<span><span class="co">#&gt; [22] textshaping_0.3.6      cli_3.3.0              Biobase_2.56.0        </span></span>
<span><span class="co">#&gt; [25] desc_1.4.2             labeling_0.4.2         sass_0.4.2            </span></span>
<span><span class="co">#&gt; [28] scales_1.2.1           pkgdown_2.0.6          systemfonts_1.0.4     </span></span>
<span><span class="co">#&gt; [31] stringr_1.4.1          digest_0.6.29          rmarkdown_2.16        </span></span>
<span><span class="co">#&gt; [34] XVector_0.36.0         pkgconfig_2.0.3        htmltools_0.5.3       </span></span>
<span><span class="co">#&gt; [37] highr_0.9              fastmap_1.1.0          rlang_1.0.5           </span></span>
<span><span class="co">#&gt; [40] rstudioapi_0.14        farver_2.1.1           jquerylib_0.1.4       </span></span>
<span><span class="co">#&gt; [43] generics_0.1.3         jsonlite_1.8.0         car_3.1-0             </span></span>
<span><span class="co">#&gt; [46] RCurl_1.98-1.6         magrittr_2.0.3         GenomeInfoDbData_1.2.8</span></span>
<span><span class="co">#&gt; [49] biomformat_1.24.0      Matrix_1.5-1           Rcpp_1.0.8.3          </span></span>
<span><span class="co">#&gt; [52] munsell_0.5.0          S4Vectors_0.34.0       Rhdf5lib_1.18.2       </span></span>
<span><span class="co">#&gt; [55] fansi_1.0.3            abind_1.4-5            ape_5.6-2             </span></span>
<span><span class="co">#&gt; [58] lifecycle_1.0.2        stringi_1.7.6          yaml_2.3.5            </span></span>
<span><span class="co">#&gt; [61] carData_3.0-5          MASS_7.3-57            zlibbioc_1.42.0       </span></span>
<span><span class="co">#&gt; [64] rhdf5_2.40.0           Rtsne_0.16             plyr_1.8.7            </span></span>
<span><span class="co">#&gt; [67] grid_4.2.1             parallel_4.2.1         crayon_1.5.1          </span></span>
<span><span class="co">#&gt; [70] lattice_0.20-45        Biostrings_2.64.0      splines_4.2.1         </span></span>
<span><span class="co">#&gt; [73] multtest_2.52.0        pillar_1.8.1           igraph_1.3.1          </span></span>
<span><span class="co">#&gt; [76] ggsignif_0.6.3         reshape2_1.4.4         codetools_0.2-18      </span></span>
<span><span class="co">#&gt; [79] stats4_4.2.1           glue_1.6.2             evaluate_0.16         </span></span>
<span><span class="co">#&gt; [82] data.table_1.14.2      vctrs_0.4.1            foreach_1.5.2         </span></span>
<span><span class="co">#&gt; [85] gtable_0.3.1           purrr_0.3.4            tidyr_1.2.0           </span></span>
<span><span class="co">#&gt; [88] assertthat_0.2.1       cachem_1.0.6           xfun_0.31             </span></span>
<span><span class="co">#&gt; [91] broom_1.0.1            rstatix_0.7.0          ragg_1.2.2            </span></span>
<span><span class="co">#&gt; [94] survival_3.3-1         iterators_1.0.14       memoise_2.0.1         </span></span>
<span><span class="co">#&gt; [97] IRanges_2.30.0         cluster_2.1.3          ellipsis_0.3.2</span></span></code></pre></div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Sudarshan Shetty, Leo Lahti.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.6.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
